{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/run/media/nazif/2F946E411BA61D49/thesis\n"
     ]
    }
   ],
   "source": [
    "%cd ..\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "from scripts.utils_latest import *\n",
    "\n",
    "pd.options.mode.chained_assignment = None  # Disable the SettingWithCopyWarning warning globally\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>339439</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>339608</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>347721</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16</td>\n",
       "      <td>360070</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>396146</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr     pos ref alt  driver\n",
       "0  16  339439   C   T       1\n",
       "1  16  339608   C   T       1\n",
       "2  16  347721   C   T       1\n",
       "3  16  360070   C   A       1\n",
       "4  16  396146   A   T       1"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"sana/dataset.csv\")\n",
    "\n",
    "df[\"ref\"] = df[\"ref\"].replace(\"-\", \"\")\n",
    "df[\"alt\"] = df[\"alt\"].replace(\"-\", \"\")\n",
    "df.rename(columns={\"start\": \"pos\"}, inplace=True)\n",
    "df.loc[df['chr'] == 'Un_KI270742v1', 'chr'] = '1'\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.ncbi.nlm.nih.gov/clinvar/RCV000140032/ this page shows that this variant chromosome \"Un_KI270742v1\" is at Chr1: 142728921 - 142907112 (on Assembly GRCh37)\n",
    "\n",
    "so I added the line \n",
    "\n",
    "```df.loc[df['chr'] == 'Un_KI270742v1', 'chr'] = '1'```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add an 'id' column by combining 'chr', 'pos', 'ref', and 'alt' columns\n",
    "df['id'] = df['chr'] + '_' + df['pos'].astype(str) + '_' + df['ref'] + '_' + df['alt']\n",
    "\n",
    "# preparing mutation len values\n",
    "df.loc[:, \"ref_len\"] = df[\"ref\"].str.len()\n",
    "df.loc[:, \"alt_len\"] = df[\"alt\"].str.len()\n",
    "\n",
    "# fetching sequences from fasta files to validate\n",
    "\n",
    "# if affected sequences are longer than 1, fetch nucleotides at the specified interval\n",
    "df.loc[df['ref_len'] > 1, 'fetched_nucleotides'] = df.apply(lambda x: get_nucleotides_in_interval(x['chr'], x['pos'], x[\"pos\"]+x[\"ref_len\"]-1), axis=1)\n",
    "# if affected sequences are equal to 1, fetch nucleotides at the specified pos\n",
    "df.loc[df['ref_len'] == 1, 'fetched_nucleotides'] = df.apply(lambda x: get_nucleotide_at_position(x['chr'], x['pos']), axis=1)\n",
    "# same but for 0 length\n",
    "df.loc[df['ref_len'] == 0, 'fetched_nucleotides'] = \"\"\n",
    "\n",
    "# Add an 'is_nucleotides_same' column to check if fetched nucleotides are the same as 'ref'\n",
    "df.loc[:, 'is_nucleotides_same'] = df[\"fetched_nucleotides\"] == df[\"ref\"]\n",
    "\n",
    "# preparing reference indices\n",
    "df.loc[:, \"ref_start\"] = df[\"pos\"]\n",
    "df.loc[:, \"ref_end\"] = df[\"pos\"] + df[\"ref_len\"]\n",
    "\n",
    "# preparing extended sequences\n",
    "df.loc[:, 'upstream_sequence'] = df.apply(lambda x: get_nucleotides_in_interval(x['chr'], x['ref_start']-30, x[\"ref_start\"]-1), axis=1)\n",
    "df.loc[:, 'downstream_sequence'] = df.apply(lambda x: get_nucleotides_in_interval(x['chr'], x['ref_end'], x[\"ref_end\"]+29), axis=1)\n",
    "\n",
    "# preparing unprocessed sequence\n",
    "df.loc[:,'sequence'] = df[\"upstream_sequence\"] + df[\"ref\"] + df[\"downstream_sequence\"]\n",
    "\n",
    "\n",
    "# preparing mutated sequence\n",
    "df.loc[:,'mutated_sequence'] = df[\"upstream_sequence\"] + df[\"alt\"] + df[\"downstream_sequence\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>driver</th>\n",
       "      <th>id</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>alt_len</th>\n",
       "      <th>fetched_nucleotides</th>\n",
       "      <th>is_nucleotides_same</th>\n",
       "      <th>ref_start</th>\n",
       "      <th>ref_end</th>\n",
       "      <th>upstream_sequence</th>\n",
       "      <th>downstream_sequence</th>\n",
       "      <th>sequence</th>\n",
       "      <th>mutated_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16</td>\n",
       "      <td>339439</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>16_339439_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>339439</td>\n",
       "      <td>339440</td>\n",
       "      <td>GGGGAGGGGGCACCCCAGCCCTCACACTCA</td>\n",
       "      <td>CTGTAGCTGCCCTTTTTGGTCAGCAGCTCC</td>\n",
       "      <td>GGGGAGGGGGCACCCCAGCCCTCACACTCACCTGTAGCTGCCCTTT...</td>\n",
       "      <td>GGGGAGGGGGCACCCCAGCCCTCACACTCATCTGTAGCTGCCCTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>339608</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>16_339608_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>339608</td>\n",
       "      <td>339609</td>\n",
       "      <td>CCGCCGCCCACCTTCCTCTGCGATCTTGTC</td>\n",
       "      <td>TGGGGAAAGAGATGCAGCGGTGGTACCTGG</td>\n",
       "      <td>CCGCCGCCCACCTTCCTCTGCGATCTTGTCCTGGGGAAAGAGATGC...</td>\n",
       "      <td>CCGCCGCCCACCTTCCTCTGCGATCTTGTCTTGGGGAAAGAGATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16</td>\n",
       "      <td>347721</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>16_347721_C_T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>347721</td>\n",
       "      <td>347722</td>\n",
       "      <td>GCCCCCTCCTCACTGACAGGCGCACGCTCA</td>\n",
       "      <td>CTGTGGGCGAGGCCATCACTGGCGTTGGGG</td>\n",
       "      <td>GCCCCCTCCTCACTGACAGGCGCACGCTCACCTGTGGGCGAGGCCA...</td>\n",
       "      <td>GCCCCCTCCTCACTGACAGGCGCACGCTCATCTGTGGGCGAGGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16</td>\n",
       "      <td>360070</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>16_360070_C_A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>360070</td>\n",
       "      <td>360071</td>\n",
       "      <td>TTACGGATCCTGTATGGGGGGATCCCATCC</td>\n",
       "      <td>TGTCCAGGAGAAAGAGGCAGCCGTTAACTC</td>\n",
       "      <td>TTACGGATCCTGTATGGGGGGATCCCATCCCTGTCCAGGAGAAAGA...</td>\n",
       "      <td>TTACGGATCCTGTATGGGGGGATCCCATCCATGTCCAGGAGAAAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>396146</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>16_396146_A_T</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>396146</td>\n",
       "      <td>396147</td>\n",
       "      <td>TACAGAAAGTGGACGCCTGGCGTCGGACTC</td>\n",
       "      <td>CCTGAACTCTCTGCCTTCGCTGTACCGTCT</td>\n",
       "      <td>TACAGAAAGTGGACGCCTGGCGTCGGACTCACCTGAACTCTCTGCC...</td>\n",
       "      <td>TACAGAAAGTGGACGCCTGGCGTCGGACTCTCCTGAACTCTCTGCC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr     pos ref alt  driver             id  ref_len  alt_len  \\\n",
       "0  16  339439   C   T       1  16_339439_C_T        1        1   \n",
       "1  16  339608   C   T       1  16_339608_C_T        1        1   \n",
       "2  16  347721   C   T       1  16_347721_C_T        1        1   \n",
       "3  16  360070   C   A       1  16_360070_C_A        1        1   \n",
       "4  16  396146   A   T       1  16_396146_A_T        1        1   \n",
       "\n",
       "  fetched_nucleotides  is_nucleotides_same  ref_start  ref_end  \\\n",
       "0                   C                 True     339439   339440   \n",
       "1                   C                 True     339608   339609   \n",
       "2                   C                 True     347721   347722   \n",
       "3                   C                 True     360070   360071   \n",
       "4                   A                 True     396146   396147   \n",
       "\n",
       "                upstream_sequence             downstream_sequence  \\\n",
       "0  GGGGAGGGGGCACCCCAGCCCTCACACTCA  CTGTAGCTGCCCTTTTTGGTCAGCAGCTCC   \n",
       "1  CCGCCGCCCACCTTCCTCTGCGATCTTGTC  TGGGGAAAGAGATGCAGCGGTGGTACCTGG   \n",
       "2  GCCCCCTCCTCACTGACAGGCGCACGCTCA  CTGTGGGCGAGGCCATCACTGGCGTTGGGG   \n",
       "3  TTACGGATCCTGTATGGGGGGATCCCATCC  TGTCCAGGAGAAAGAGGCAGCCGTTAACTC   \n",
       "4  TACAGAAAGTGGACGCCTGGCGTCGGACTC  CCTGAACTCTCTGCCTTCGCTGTACCGTCT   \n",
       "\n",
       "                                            sequence  \\\n",
       "0  GGGGAGGGGGCACCCCAGCCCTCACACTCACCTGTAGCTGCCCTTT...   \n",
       "1  CCGCCGCCCACCTTCCTCTGCGATCTTGTCCTGGGGAAAGAGATGC...   \n",
       "2  GCCCCCTCCTCACTGACAGGCGCACGCTCACCTGTGGGCGAGGCCA...   \n",
       "3  TTACGGATCCTGTATGGGGGGATCCCATCCCTGTCCAGGAGAAAGA...   \n",
       "4  TACAGAAAGTGGACGCCTGGCGTCGGACTCACCTGAACTCTCTGCC...   \n",
       "\n",
       "                                    mutated_sequence  \n",
       "0  GGGGAGGGGGCACCCCAGCCCTCACACTCATCTGTAGCTGCCCTTT...  \n",
       "1  CCGCCGCCCACCTTCCTCTGCGATCTTGTCTTGGGGAAAGAGATGC...  \n",
       "2  GCCCCCTCCTCACTGACAGGCGCACGCTCATCTGTGGGCGAGGCCA...  \n",
       "3  TTACGGATCCTGTATGGGGGGATCCCATCCATGTCCAGGAGAAAGA...  \n",
       "4  TACAGAAAGTGGACGCCTGGCGTCGGACTCTCCTGAACTCTCTGCC...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.is_nucleotides_same.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# adding is_mirna column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_is_mirna_column(df, grch=38):\n",
    "    coords = pd.read_csv(f\"data/processed/mirbase/mirna_coordinates_grch{grch}/coordinates.csv\")\n",
    "    df['is_mirna'] = 0\n",
    "\n",
    "    # Iterate over each mutation in the mutations dataframe\n",
    "    for index, row in df.iterrows():\n",
    "        mutation_chr = row['chr']\n",
    "        mutation_start = row['pos']\n",
    "\n",
    "        # Check if the mutation falls into any of the RNAs\n",
    "        matching_rnas = coords[(coords['chr'] == mutation_chr) & (coords['start'] <= mutation_start) & (coords['end'] >= mutation_start)]\n",
    "\n",
    "        if not matching_rnas.empty:\n",
    "            # Update the 'is_mirna' column to 1 for the current mutation\n",
    "            df.at[index, 'is_mirna'] = 1\n",
    "            \n",
    "    return df\n",
    "\n",
    "df = generate_is_mirna_column(df, grch=38)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# finding matches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = find_matches_for_vcfs(df[:2])\n",
    "results_mutated = find_matches_for_vcfs(df[:2], mutated=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results[\"is_mutated\"] = 0\n",
    "results_mutated[\"is_mutated\"] = 1\n",
    "final =pd.concat([results, results_mutated], axis=0)\n",
    "\n",
    "final.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final.to_csv(\"sana/initial_results.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis-4DYweHFz-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
