{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/run/media/nazif/2F946E411BA61D49/thesis\n"
     ]
    }
   ],
   "source": [
    "cd ../../.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl60/Homo_sapiens.GRCh37.60.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl60/Homo_sapiens.GRCh37.60.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl60/Homo_sapiens.GRCh37.60.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "\n",
    "# pyensembl\n",
    "from pyensembl import EnsemblRelease\n",
    "import os\n",
    "os.environ[\"PYENSEMBL_CACHE_DIR\"] = \"../data\"\n",
    "ens60 = EnsemblRelease(60)\n",
    "ens60.download()\n",
    "ens60.index()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# read CLASH mmc1.txt and save to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/raw/clash/mmc1.txt\") as f:\n",
    "\n",
    "    # iter(f) is the same as (line for line in f)\n",
    "    lines = iter(f)\n",
    "\n",
    "    columns = []\n",
    "    data = []\n",
    "    for line in lines:\n",
    "        if line.startswith(\"#\"):\n",
    "            line = line[1:]\n",
    "            row = next(csv.reader([line], delimiter=\"\\t\"))\n",
    "            columns.append(row)\n",
    "        else:\n",
    "            row = next(csv.reader([line], delimiter=\"\\t\"))\n",
    "            data.append(row)\n",
    "\n",
    "    # removes header text\n",
    "    columns.pop(0)\n",
    "\n",
    "    # create the output directory if it doesn\"t exist\n",
    "    output_dir = \".\"\n",
    "    os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "    # write the output CSV files\n",
    "    with open(os.path.join(output_dir, \"clash_column_details.csv\"), \"w\") as f1, \\\n",
    "            open(os.path.join(output_dir, \"clash_raw.csv\"), \"w\") as f2:\n",
    "        writer1 = csv.writer(f1, delimiter=\",\")\n",
    "        writer2 = csv.writer(f2, delimiter=\",\")\n",
    "        writer1.writerows(columns)\n",
    "        writer2.writerows(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# wrangling previously exported csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"data/raw/clash/clash_raw.csv\")\n",
    "\n",
    "# dropping unused CLASH columns\n",
    "cols_to_drop = [\"miRNA_start\", \"miRNA_end\", \"chimeras_decompressed\",\n",
    "                \"experiments\", \"experiments_list\", \"microRNA_first\", \"two_way_merged\",\n",
    "                \"log2_target_enrichment\", \"CLASH_single_reads_ovlp\",\n",
    "                \"5'UTR\", \"CDS\", \"3'UTR\", \"conservation_score\",\n",
    "                \"CLASH_cluster_ovlp\", \"PAR_CLIP_cluster_ovlp\"]\n",
    "\n",
    "df.drop(columns=cols_to_drop, inplace=True)\n",
    "\n",
    "# process microRNA_name and mRNA_name columns\n",
    "new_mirna_cols = df[\"microRNA_name\"].str.split(\"_\", expand=True)\n",
    "new_mirna_cols.columns = [\"mirna_accession\", \"temp1\", \"mirna_name\", \"temp2\"]\n",
    "df = pd.concat([df, new_mirna_cols], axis=1)\n",
    "\n",
    "new_mrna_cols = df[\"mRNA_name\"].str.split(\"_\", expand=True)\n",
    "new_mrna_cols.columns = [\"ensg\", \"enst\", \"gene_name\", \"temp3\"]\n",
    "df = pd.concat([df, new_mrna_cols], axis=1)\n",
    "\n",
    "# dropping temporary columns\n",
    "temp_cols = [\"microRNA_name\", \"mRNA_name\", \"temp1\", \"temp2\", \"temp3\"]\n",
    "df.drop(columns=temp_cols, inplace=True)\n",
    "\n",
    "# renaming columns\n",
    "rename_dict = {\n",
    "    \"seq_ID\": \"id\",\n",
    "    \"miRNA_seq\": \"mirna_sequence\",\n",
    "    \"mRNA_seq_extended\": \"mrna_sequence\",\n",
    "    \"mRNA_start\": \"true_start\",\n",
    "    \"mRNA_end_extended\": \"true_end\",\n",
    "    \"seed_type\": \"true_seed_type\",\n",
    "    \"folding_class\": \"true_folding_class\"\n",
    "}\n",
    "\n",
    "df = df.rename(columns=rename_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# augmenting with pyensembl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequence_slice(row, start_offset=0, end_offset=0):\n",
    "    sequence = row[\"full_sequence\"]\n",
    "    start = row[\"true_start\"] - 1 - start_offset\n",
    "    end = row[\"true_end\"] + end_offset\n",
    "    return sequence[start:end]\n",
    "\n",
    "\n",
    "def get_upstream_n_nucleotides(row, n):\n",
    "    sequence = row[\"full_sequence\"]\n",
    "    start = row[\"true_start\"] - 1 - n\n",
    "    end = row[\"true_start\"] - 1\n",
    "    return sequence[start:end]\n",
    "\n",
    "\n",
    "def get_downstream_n_nucleotides(row, n):\n",
    "    sequence = row[\"full_sequence\"]\n",
    "    start = row[\"true_end\"]\n",
    "    end = row[\"true_end\"] + n\n",
    "    return sequence[start:end]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get unique ENSTs\n",
    "ensts_to_fetch = df.enst.unique().tolist()\n",
    "\n",
    "# get sequences of ENSTs from ENSEMBL 60 to a dict\n",
    "seq_dict = {\n",
    "    i: ens60.transcript_by_id(i).sequence\n",
    "    if ens60.transcript_by_id(i).sequence\n",
    "    else None\n",
    "    for i in ensts_to_fetch\n",
    "}\n",
    "\n",
    "# appending full sequences to the clash df\n",
    "df[\"full_sequence\"] = df[\"enst\"].map(seq_dict)\n",
    "\n",
    "df[\"upstream_30_nucleotides\"] = df.apply(\n",
    "    get_upstream_n_nucleotides, args=(30,), axis=1)\n",
    "df[\"sequence_slice\"] = df.apply(get_sequence_slice, axis=1)\n",
    "df[\"downstream_30_nucleotides\"] = df.apply(\n",
    "    get_downstream_n_nucleotides, args=(30,), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>mirna_sequence</th>\n",
       "      <th>true_start</th>\n",
       "      <th>true_end</th>\n",
       "      <th>mrna_sequence</th>\n",
       "      <th>true_seed_type</th>\n",
       "      <th>num_basepairs</th>\n",
       "      <th>seed_basepairs</th>\n",
       "      <th>folding_energy</th>\n",
       "      <th>true_folding_class</th>\n",
       "      <th>...</th>\n",
       "      <th>ensg</th>\n",
       "      <th>enst</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>full_sequence</th>\n",
       "      <th>upstream_30_nucleotides</th>\n",
       "      <th>sequence_slice</th>\n",
       "      <th>downstream_30_nucleotides</th>\n",
       "      <th>transcript_chr</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0727A-1038930_1</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>1791</td>\n",
       "      <td>1890</td>\n",
       "      <td>ATTTGTATCTACGATAAAAATTTTTATACAGAACCTACTGCCTCAA...</td>\n",
       "      <td>noncanonical_seed</td>\n",
       "      <td>20</td>\n",
       "      <td>6</td>\n",
       "      <td>-25.1</td>\n",
       "      <td>III</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000113328</td>\n",
       "      <td>ENST00000340828</td>\n",
       "      <td>CCNG1</td>\n",
       "      <td>AGGGCAGGCGCGGCCCCTTCGGCTCCGAGCTGACCCTGATCAGGGC...</td>\n",
       "      <td>TTTAATATTTTTTTCTAGAAAACAGGTGAC</td>\n",
       "      <td>ATTTGTATCTACGATAAAAATTTTTATACAGAACCTACTGCCTCAA...</td>\n",
       "      <td>TATCACTTCGAAAACTTGCTTTCCCACACT</td>\n",
       "      <td>5</td>\n",
       "      <td>162864587</td>\n",
       "      <td>162872022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1HS-1112536_1</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>3857</td>\n",
       "      <td>3928</td>\n",
       "      <td>CAGGAAATACCCGTGCAACCAACTACCTCATATTCCATTCAGAATT...</td>\n",
       "      <td>9-mer</td>\n",
       "      <td>17</td>\n",
       "      <td>6</td>\n",
       "      <td>-24.4</td>\n",
       "      <td>II</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000100697</td>\n",
       "      <td>ENST00000343455</td>\n",
       "      <td>DICER1</td>\n",
       "      <td>CGGAGGCGCGGCGCAGGCTGCTGCAGGCCCAGGTGAATGGAGTAAC...</td>\n",
       "      <td>TGCCAAGGAAATCAGCTAAATTACTACAAG</td>\n",
       "      <td>CAGGAAATACCCGTGCAACCAACTACCTCATATTCCATTCAGAATT...</td>\n",
       "      <td>CCCAGCGATGAATGTACTCTCCTGAGTAAT</td>\n",
       "      <td>14</td>\n",
       "      <td>95552566</td>\n",
       "      <td>95623759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L2HS-818542_2</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>2385</td>\n",
       "      <td>2434</td>\n",
       "      <td>ACCCGCTATATGACCTGATGCCTTTCCTTCATTAAAGATGATTCTG...</td>\n",
       "      <td>noncanonical_seed</td>\n",
       "      <td>19</td>\n",
       "      <td>6</td>\n",
       "      <td>-22.2</td>\n",
       "      <td>III</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000080546</td>\n",
       "      <td>ENST00000436639</td>\n",
       "      <td>SESN1</td>\n",
       "      <td>GATTGCCAGGGCCGCCCTGTGCCCTCTGGCTCGGCGGTGGTGGGCG...</td>\n",
       "      <td>GCAGAACTCCTTTATGCTCTGAGAGCCATT</td>\n",
       "      <td>ACCCGCTATATGACCTGATGCCTTTCCTTCATTAAAGATGATTCTG...</td>\n",
       "      <td>GATCAGCAGATATAGTCTACAAGGGGGAAG</td>\n",
       "      <td>6</td>\n",
       "      <td>109307640</td>\n",
       "      <td>109416022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L2HS-1161339_2</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>6570</td>\n",
       "      <td>6623</td>\n",
       "      <td>CAATGACTATGCAACCATACCTTACCACTAAATGTAGTACGCAAAA...</td>\n",
       "      <td>noncanonical_seed</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>-22.1</td>\n",
       "      <td>III</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000164190</td>\n",
       "      <td>ENST00000282516</td>\n",
       "      <td>NIPBL</td>\n",
       "      <td>TCCGGTCGGCATTTTGTTCTGAGAGGGAGAGACGGAACGAGAGAGA...</td>\n",
       "      <td>AAATAAGACCCCAGCTCATGGTTAAACATG</td>\n",
       "      <td>CAATGACTATGCAACCATACCTTACCACTAAATGTAGTACGCAAAA...</td>\n",
       "      <td>TGGTTATCTGCAATGTTGCAAAAATCCTAG</td>\n",
       "      <td>5</td>\n",
       "      <td>36876861</td>\n",
       "      <td>37066515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L2-407944_2</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>1164</td>\n",
       "      <td>1208</td>\n",
       "      <td>AATACTGGGAAACCTGCTACTTCGTCAGCTAACCAGAAACCTGTG</td>\n",
       "      <td>noncanonical_seed</td>\n",
       "      <td>18</td>\n",
       "      <td>6</td>\n",
       "      <td>-21.9</td>\n",
       "      <td>III</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000138785</td>\n",
       "      <td>ENST00000340139</td>\n",
       "      <td>INTS12</td>\n",
       "      <td>AGGGACCACCGGGAACAGACGGATCGGCAGGGCGGGGCGGAACGGT...</td>\n",
       "      <td>ACAGCAAAATTGAGTTCAACAACACAAAAC</td>\n",
       "      <td>AATACTGGGAAACCTGCTACTTCGTCAGCTAACCAGAAACCTGTG</td>\n",
       "      <td>GGTTTGACTGGTCTGGCAACATCATCCAAA</td>\n",
       "      <td>4</td>\n",
       "      <td>106603784</td>\n",
       "      <td>106629838</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id          mirna_sequence  true_start  true_end  \\\n",
       "0  0727A-1038930_1  TGAGGTAGTAGGTTGTATAGTT        1791      1890   \n",
       "1   L1HS-1112536_1  TGAGGTAGTAGGTTGTATAGTT        3857      3928   \n",
       "2    L2HS-818542_2  TGAGGTAGTAGGTTGTATAGTT        2385      2434   \n",
       "3   L2HS-1161339_2  TGAGGTAGTAGGTTGTATAGTT        6570      6623   \n",
       "4      L2-407944_2  TGAGGTAGTAGGTTGTATAGTT        1164      1208   \n",
       "\n",
       "                                       mrna_sequence     true_seed_type  \\\n",
       "0  ATTTGTATCTACGATAAAAATTTTTATACAGAACCTACTGCCTCAA...  noncanonical_seed   \n",
       "1  CAGGAAATACCCGTGCAACCAACTACCTCATATTCCATTCAGAATT...              9-mer   \n",
       "2  ACCCGCTATATGACCTGATGCCTTTCCTTCATTAAAGATGATTCTG...  noncanonical_seed   \n",
       "3  CAATGACTATGCAACCATACCTTACCACTAAATGTAGTACGCAAAA...  noncanonical_seed   \n",
       "4      AATACTGGGAAACCTGCTACTTCGTCAGCTAACCAGAAACCTGTG  noncanonical_seed   \n",
       "\n",
       "   num_basepairs  seed_basepairs  folding_energy true_folding_class  ...  \\\n",
       "0             20               6           -25.1                III  ...   \n",
       "1             17               6           -24.4                 II  ...   \n",
       "2             19               6           -22.2                III  ...   \n",
       "3             18               6           -22.1                III  ...   \n",
       "4             18               6           -21.9                III  ...   \n",
       "\n",
       "              ensg             enst gene_name  \\\n",
       "0  ENSG00000113328  ENST00000340828     CCNG1   \n",
       "1  ENSG00000100697  ENST00000343455    DICER1   \n",
       "2  ENSG00000080546  ENST00000436639     SESN1   \n",
       "3  ENSG00000164190  ENST00000282516     NIPBL   \n",
       "4  ENSG00000138785  ENST00000340139    INTS12   \n",
       "\n",
       "                                       full_sequence  \\\n",
       "0  AGGGCAGGCGCGGCCCCTTCGGCTCCGAGCTGACCCTGATCAGGGC...   \n",
       "1  CGGAGGCGCGGCGCAGGCTGCTGCAGGCCCAGGTGAATGGAGTAAC...   \n",
       "2  GATTGCCAGGGCCGCCCTGTGCCCTCTGGCTCGGCGGTGGTGGGCG...   \n",
       "3  TCCGGTCGGCATTTTGTTCTGAGAGGGAGAGACGGAACGAGAGAGA...   \n",
       "4  AGGGACCACCGGGAACAGACGGATCGGCAGGGCGGGGCGGAACGGT...   \n",
       "\n",
       "          upstream_30_nucleotides  \\\n",
       "0  TTTAATATTTTTTTCTAGAAAACAGGTGAC   \n",
       "1  TGCCAAGGAAATCAGCTAAATTACTACAAG   \n",
       "2  GCAGAACTCCTTTATGCTCTGAGAGCCATT   \n",
       "3  AAATAAGACCCCAGCTCATGGTTAAACATG   \n",
       "4  ACAGCAAAATTGAGTTCAACAACACAAAAC   \n",
       "\n",
       "                                      sequence_slice  \\\n",
       "0  ATTTGTATCTACGATAAAAATTTTTATACAGAACCTACTGCCTCAA...   \n",
       "1  CAGGAAATACCCGTGCAACCAACTACCTCATATTCCATTCAGAATT...   \n",
       "2  ACCCGCTATATGACCTGATGCCTTTCCTTCATTAAAGATGATTCTG...   \n",
       "3  CAATGACTATGCAACCATACCTTACCACTAAATGTAGTACGCAAAA...   \n",
       "4      AATACTGGGAAACCTGCTACTTCGTCAGCTAACCAGAAACCTGTG   \n",
       "\n",
       "        downstream_30_nucleotides transcript_chr transcript_start  \\\n",
       "0  TATCACTTCGAAAACTTGCTTTCCCACACT              5        162864587   \n",
       "1  CCCAGCGATGAATGTACTCTCCTGAGTAAT             14         95552566   \n",
       "2  GATCAGCAGATATAGTCTACAAGGGGGAAG              6        109307640   \n",
       "3  TGGTTATCTGCAATGTTGCAAAAATCCTAG              5         36876861   \n",
       "4  GGTTTGACTGGTCTGGCAACATCATCCAAA              4        106603784   \n",
       "\n",
       "  transcript_end  \n",
       "0      162872022  \n",
       "1       95623759  \n",
       "2      109416022  \n",
       "3       37066515  \n",
       "4      106629838  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ensts = df[\"enst\"].values.tolist()\n",
    "\n",
    "chromosomes = []\n",
    "starts = []\n",
    "ends = []\n",
    "strands = []\n",
    "ensembl_sequences = []\n",
    "\n",
    "\n",
    "for i in ensts:\n",
    "    obj = ens60.transcript_by_id(i)\n",
    "\n",
    "    chromosomes.append(obj.contig)\n",
    "    starts.append(obj.start)\n",
    "    ends.append(obj.end)\n",
    "    strands.append(obj.strand)\n",
    "    ensembl_sequences.append(obj.sequence)\n",
    "\n",
    "df[\"transcript_chr\"] = chromosomes\n",
    "df[\"transcript_start\"] = starts\n",
    "df[\"transcript_end\"] = ends\n",
    "\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"data/processed/clash/clash_parsed.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
