{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PREPROCESS IS WORKING \n",
    "\n",
    "## converting coords into grch38 has problems\n",
    "\n",
    "## .bed file has 23k lines while our original data had 18k rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_clash_data(filename=\"mmc1.txt\"):\n",
    "    \"\"\"\n",
    "    Parse Clash data from the given file and write the results to output files.\n",
    "\n",
    "    Args:\n",
    "        filename (str): The path to the input file. Default is \"data/raw/clash/mmc1.txt\".\n",
    "\n",
    "    Raises:\n",
    "        FileNotFoundError: If the input file does not exist.\n",
    "\n",
    "    Returns:\n",
    "        None.\n",
    "    \"\"\"\n",
    "\n",
    "    with open(filename) as f:\n",
    "        \n",
    "        # iter(f) is the same as (line for line in f)\n",
    "        lines = iter(f)\n",
    "\n",
    "        columns = []\n",
    "        data = []\n",
    "        for line in lines:\n",
    "            if line.startswith(\"#\"):\n",
    "                line = line[1:]\n",
    "                row = next(csv.reader([line], delimiter=\"\\t\"))\n",
    "                columns.append(row)\n",
    "            else:\n",
    "                row = next(csv.reader([line], delimiter=\"\\t\"))\n",
    "                data.append(row)\n",
    "\n",
    "        # removes header text\n",
    "        columns.pop(0)\n",
    "\n",
    "        # create the output directory if it doesn\"t exist\n",
    "        output_dir = \".\"\n",
    "        os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "        # write the output CSV files\n",
    "        with open(os.path.join(output_dir, \"clash_column_details.csv\"), \"w\") as f1, \\\n",
    "             open(os.path.join(output_dir, \"clash_raw.csv\"), \"w\") as f2:\n",
    "            writer1 = csv.writer(f1, delimiter=\",\")\n",
    "            writer2 = csv.writer(f2, delimiter=\",\")\n",
    "            writer1.writerows(columns)\n",
    "            writer2.writerows(data)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## write to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_clash_data()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## processing miRNA and mRNA names and saving to csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"clash_raw.csv\")\n",
    "\n",
    "# process microRNA_name column\n",
    "new_cols = df[\"microRNA_name\"].str.split(\"_\", expand=True)\n",
    "new_cols.columns = [\"mirna_accession\", \"from\", \"mirna_name\", \"temp\"]\n",
    "df = pd.concat([df, new_cols], axis=1)\n",
    "df = df.drop(\"microRNA_name\", axis=1)\n",
    "df = df.drop(\"temp\", axis=1)\n",
    "df = df.drop(\"from\", axis=1)\n",
    "\n",
    "# process mRNA_name column\n",
    "new_cols = df[\"mRNA_name\"].str.split(\"_\", expand=True)\n",
    "new_cols.columns = [\"ensg\", \"enst\", \"gene_name\", \"temp\"]\n",
    "df = pd.concat([df, new_cols], axis=1)\n",
    "df = df.drop(\"mRNA_name\", axis=1)\n",
    "df = df.drop(\"temp\", axis=1)\n",
    "\n",
    "df.to_csv(\"clash_raw_renamed_columns.csv\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# parsing clash_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_clash_raw(drop_irrelevant_columns=True):\n",
    "\n",
    "    df = pd.read_csv(\"clash_raw.csv\")\n",
    "\n",
    "    if drop_irrelevant_columns:\n",
    "\n",
    "        cols_to_drop = ['miRNA_start', 'miRNA_end','chimeras_decompressed',\n",
    "                        'experiments', 'experiments_list', 'microRNA_first', 'two_way_merged',\n",
    "                        \"5'UTR\", 'CDS', \"3'UTR\", 'conservation_score',\n",
    "                        'log2_target_enrichment', 'CLASH_single_reads_ovlp',\n",
    "                        'CLASH_cluster_ovlp', 'PAR_CLIP_cluster_ovlp']\n",
    "\n",
    "        df.drop(columns=cols_to_drop, inplace=True)\n",
    "\n",
    "\n",
    "    # process microRNA_name column\n",
    "    new_cols = df[\"microRNA_name\"].str.split(\"_\", expand=True)\n",
    "    new_cols.columns = [\"mirna_accession\", \"from\", \"mirna_name\", \"temp\"]\n",
    "    df = pd.concat([df, new_cols], axis=1)\n",
    "    df = df.drop(\"microRNA_name\", axis=1)\n",
    "    df = df.drop(\"temp\", axis=1)\n",
    "    df = df.drop(\"from\", axis=1)\n",
    "\n",
    "    # process mRNA_name column\n",
    "    new_cols = df[\"mRNA_name\"].str.split(\"_\", expand=True)\n",
    "    new_cols.columns = [\"ensg\", \"enst\", \"gene_name\", \"temp\"]\n",
    "    df = pd.concat([df, new_cols], axis=1)\n",
    "    df = df.drop(\"mRNA_name\", axis=1)\n",
    "    df = df.drop(\"temp\", axis=1)\n",
    "    \n",
    "    # dropping rows where len(mrna_sequence) < len(mirna_sequence)\n",
    "    df = df[df.mRNA_seq_extended.str.len() >= df.miRNA_seq.str.len()]\n",
    "    \n",
    "    rename_dict = {\n",
    "    \"seq_ID\": \"id\",\n",
    "    \"miRNA_seq\": \"mirna_sequence\",\n",
    "    \"mRNA_seq_extended\": \"mrna_sequence\",\n",
    "    \"mRNA_start\": \"true_start\",\n",
    "    \"mRNA_end_extended\": \"true_end\",\n",
    "    \"seed_type\": \"true_seed_type\",\n",
    "    \"folding_class\": \"true_folding_class\"\n",
    "}\n",
    "\n",
    "    return df.rename(columns=rename_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = parse_clash_raw(drop_irrelevant_columns=True)\n",
    "df.to_csv(\"../../processed/clash/clash_parsed.csv\", index=False)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ENST into genomic coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl60/Homo_sapiens.GRCh37.60.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl60/Homo_sapiens.GRCh37.60.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl60/Homo_sapiens.GRCh37.60.pep.all.fa.gz.pickle\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>miRNA_start</th>\n",
       "      <th>miRNA_end</th>\n",
       "      <th>miRNA_seq</th>\n",
       "      <th>mRNA_start</th>\n",
       "      <th>mRNA_end_extended</th>\n",
       "      <th>mRNA_seq_extended</th>\n",
       "      <th>chimeras_decompressed</th>\n",
       "      <th>experiments</th>\n",
       "      <th>experiments_list</th>\n",
       "      <th>...</th>\n",
       "      <th>CLASH_cluster_ovlp</th>\n",
       "      <th>PAR_CLIP_cluster_ovlp</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>ensg</th>\n",
       "      <th>enst</th>\n",
       "      <th>transcript_chromosome</th>\n",
       "      <th>transcript_start</th>\n",
       "      <th>transcript_end</th>\n",
       "      <th>transcript_strand</th>\n",
       "      <th>transcript_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0727A-1038930_1</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>1791</td>\n",
       "      <td>1890</td>\n",
       "      <td>ATTTGTATCTACGATAAAAATTTTTATACAGAACCTACTGCCTCAA...</td>\n",
       "      <td>31</td>\n",
       "      <td>3</td>\n",
       "      <td>E2,E3,E4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>let-7a</td>\n",
       "      <td>ENSG00000113328</td>\n",
       "      <td>ENST00000340828</td>\n",
       "      <td>5</td>\n",
       "      <td>162864587</td>\n",
       "      <td>162872022</td>\n",
       "      <td>+</td>\n",
       "      <td>AGGGCAGGCGCGGCCCCTTCGGCTCCGAGCTGACCCTGATCAGGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>L1HS-1112536_1</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>3857</td>\n",
       "      <td>3928</td>\n",
       "      <td>CAGGAAATACCCGTGCAACCAACTACCTCATATTCCATTCAGAATT...</td>\n",
       "      <td>9</td>\n",
       "      <td>2</td>\n",
       "      <td>E3,E4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>let-7a</td>\n",
       "      <td>ENSG00000100697</td>\n",
       "      <td>ENST00000343455</td>\n",
       "      <td>14</td>\n",
       "      <td>95552566</td>\n",
       "      <td>95623759</td>\n",
       "      <td>-</td>\n",
       "      <td>CGGAGGCGCGGCGCAGGCTGCTGCAGGCCCAGGTGAATGGAGTAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>L2HS-818542_2</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>2385</td>\n",
       "      <td>2434</td>\n",
       "      <td>ACCCGCTATATGACCTGATGCCTTTCCTTCATTAAAGATGATTCTG...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>E4</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>let-7a</td>\n",
       "      <td>ENSG00000080546</td>\n",
       "      <td>ENST00000436639</td>\n",
       "      <td>6</td>\n",
       "      <td>109307640</td>\n",
       "      <td>109416022</td>\n",
       "      <td>-</td>\n",
       "      <td>GATTGCCAGGGCCGCCCTGTGCCCTCTGGCTCGGCGGTGGTGGGCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>L2HS-1161339_2</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>6570</td>\n",
       "      <td>6623</td>\n",
       "      <td>CAATGACTATGCAACCATACCTTACCACTAAATGTAGTACGCAAAA...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>E4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>let-7a</td>\n",
       "      <td>ENSG00000164190</td>\n",
       "      <td>ENST00000282516</td>\n",
       "      <td>5</td>\n",
       "      <td>36876861</td>\n",
       "      <td>37066515</td>\n",
       "      <td>+</td>\n",
       "      <td>TCCGGTCGGCATTTTGTTCTGAGAGGGAGAGACGGAACGAGAGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>L2-407944_2</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>1164</td>\n",
       "      <td>1208</td>\n",
       "      <td>AATACTGGGAAACCTGCTACTTCGTCAGCTAACCAGAAACCTGTG</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>E4</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>let-7a</td>\n",
       "      <td>ENSG00000138785</td>\n",
       "      <td>ENST00000340139</td>\n",
       "      <td>4</td>\n",
       "      <td>106603784</td>\n",
       "      <td>106629838</td>\n",
       "      <td>-</td>\n",
       "      <td>AGGGACCACCGGGAACAGACGGATCGGCAGGGCGGGGCGGAACGGT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            seq_ID  miRNA_start  miRNA_end               miRNA_seq  \\\n",
       "0  0727A-1038930_1            1         22  TGAGGTAGTAGGTTGTATAGTT   \n",
       "1   L1HS-1112536_1            1         22  TGAGGTAGTAGGTTGTATAGTT   \n",
       "2    L2HS-818542_2            1         22  TGAGGTAGTAGGTTGTATAGTT   \n",
       "3   L2HS-1161339_2            1         22  TGAGGTAGTAGGTTGTATAGTT   \n",
       "4      L2-407944_2            1         22  TGAGGTAGTAGGTTGTATAGTT   \n",
       "\n",
       "   mRNA_start  mRNA_end_extended  \\\n",
       "0        1791               1890   \n",
       "1        3857               3928   \n",
       "2        2385               2434   \n",
       "3        6570               6623   \n",
       "4        1164               1208   \n",
       "\n",
       "                                   mRNA_seq_extended  chimeras_decompressed  \\\n",
       "0  ATTTGTATCTACGATAAAAATTTTTATACAGAACCTACTGCCTCAA...                     31   \n",
       "1  CAGGAAATACCCGTGCAACCAACTACCTCATATTCCATTCAGAATT...                      9   \n",
       "2  ACCCGCTATATGACCTGATGCCTTTCCTTCATTAAAGATGATTCTG...                      2   \n",
       "3  CAATGACTATGCAACCATACCTTACCACTAAATGTAGTACGCAAAA...                      2   \n",
       "4      AATACTGGGAAACCTGCTACTTCGTCAGCTAACCAGAAACCTGTG                      2   \n",
       "\n",
       "   experiments experiments_list  ...  CLASH_cluster_ovlp  \\\n",
       "0            3         E2,E3,E4  ...                 NaN   \n",
       "1            2            E3,E4  ...                 NaN   \n",
       "2            1               E4  ...                 1.0   \n",
       "3            1               E4  ...                 NaN   \n",
       "4            1               E4  ...                 NaN   \n",
       "\n",
       "   PAR_CLIP_cluster_ovlp mirna_name             ensg             enst  \\\n",
       "0                    NaN     let-7a  ENSG00000113328  ENST00000340828   \n",
       "1                    1.0     let-7a  ENSG00000100697  ENST00000343455   \n",
       "2                    1.0     let-7a  ENSG00000080546  ENST00000436639   \n",
       "3                    NaN     let-7a  ENSG00000164190  ENST00000282516   \n",
       "4                    NaN     let-7a  ENSG00000138785  ENST00000340139   \n",
       "\n",
       "   transcript_chromosome  transcript_start  transcript_end  transcript_strand  \\\n",
       "0                      5         162864587       162872022                  +   \n",
       "1                     14          95552566        95623759                  -   \n",
       "2                      6         109307640       109416022                  -   \n",
       "3                      5          36876861        37066515                  +   \n",
       "4                      4         106603784       106629838                  -   \n",
       "\n",
       "                                 transcript_sequence  \n",
       "0  AGGGCAGGCGCGGCCCCTTCGGCTCCGAGCTGACCCTGATCAGGGC...  \n",
       "1  CGGAGGCGCGGCGCAGGCTGCTGCAGGCCCAGGTGAATGGAGTAAC...  \n",
       "2  GATTGCCAGGGCCGCCCTGTGCCCTCTGGCTCGGCGGTGGTGGGCG...  \n",
       "3  TCCGGTCGGCATTTTGTTCTGAGAGGGAGAGACGGAACGAGAGAGA...  \n",
       "4  AGGGACCACCGGGAACAGACGGATCGGCAGGGCGGGGCGGAACGGT...  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transforming ENST accessions into genomic coordinates\n",
    "\n",
    "import os\n",
    "from pyensembl import EnsemblRelease\n",
    "\n",
    "# set pyensembl cache location\n",
    "os.environ[\"PYENSEMBL_CACHE_DIR\"] = \"../../../../data\"\n",
    "# init database\n",
    "data = EnsemblRelease(60)\n",
    "# download and index db\n",
    "data.download()\n",
    "data.index()\n",
    "\n",
    "ensts = df[\"enst\"].values.tolist()\n",
    "\n",
    "chromosomes = []\n",
    "starts = []\n",
    "ends = []\n",
    "strands = []\n",
    "ensembl_sequences = []\n",
    "\n",
    "\n",
    "for i in ensts:\n",
    "    obj = data.transcript_by_id(i)\n",
    "\n",
    "    chromosomes.append(obj.contig)\n",
    "    starts.append(obj.start)\n",
    "    ends.append(obj.end)\n",
    "    strands.append(obj.strand)\n",
    "    ensembl_sequences.append(obj.sequence)\n",
    "\n",
    "df[\"transcript_chromosome\"] = chromosomes\n",
    "df[\"transcript_start\"] = starts\n",
    "df[\"transcript_end\"] = ends\n",
    "df[\"transcript_strand\"] = strands\n",
    "df[\"transcript_sequence\"] = ensembl_sequences\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1                  2055\n",
       "17                 1522\n",
       "19                 1291\n",
       "12                 1247\n",
       "16                 1019\n",
       "2                   977\n",
       "6                   923\n",
       "11                  844\n",
       "7                   841\n",
       "3                   814\n",
       "5                   803\n",
       "10                  729\n",
       "X                   728\n",
       "9                   642\n",
       "22                  579\n",
       "15                  544\n",
       "8                   517\n",
       "14                  441\n",
       "20                  434\n",
       "4                   420\n",
       "13                  258\n",
       "21                  250\n",
       "MT                  213\n",
       "18                  191\n",
       "HSCHR6_MHC_DBB      102\n",
       "HSCHR6_MHC_COX       52\n",
       "HSCHR6_MHC_APD       18\n",
       "HSCHR6_MHC_MANN      16\n",
       "HSCHR6_MHC_MCF       15\n",
       "HSCHR6_MHC_SSTO      13\n",
       "HSCHR6_MHC_QBL       11\n",
       "Y                     4\n",
       "HSCHR17_1             1\n",
       "Name: transcript_chromosome, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.transcript_chromosome.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## generating .bed file to convert old (ENSEMBL60) coordinates into most recent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment if you want to generate the bed file\n",
    "# from fuc import pybed\n",
    "\n",
    "# # filter & rename columns for pybed\n",
    "# coord_df = df[[\"chromosome\", \"start\", \"end\"]]\n",
    "# coord_df.columns = [\"Chromosome\", \"Start\", \"End\"]\n",
    "\n",
    "# # call pybed to turn into .bed file\n",
    "# bf = pybed.BedFrame.from_frame(meta=[], data=coord_df)\n",
    "# bf.to_file(\"clash_grch37_coordinates.bed\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## importing new coordinate data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop old coords\n",
    "df.drop([\"transcript_chromosome\", \"transcript_start\", \"transcript_end\"], axis=1, inplace=True)\n",
    "\n",
    "\n",
    "# read & merge new ones\n",
    "new_coords = pd.read_csv(\"../../external/clash/clash_grch38_coordinates.bed\", sep=\"\\t\", header=None)\n",
    "new_coords.columns = [\"chromosome\", \"start\", \"end\"]\n",
    "\n",
    "df = pd.concat([df, new_coords], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1                           3317\n",
       "X                           1810\n",
       "17                          1717\n",
       "19                          1494\n",
       "7                           1418\n",
       "12                          1304\n",
       "16                          1025\n",
       "2                           1022\n",
       "11                           955\n",
       "6                            949\n",
       "3                            937\n",
       "22                           926\n",
       "5                            827\n",
       "10                           759\n",
       "8                            671\n",
       "9                            659\n",
       "15                           549\n",
       "20                           548\n",
       "21                           533\n",
       "14                           530\n",
       "4                            425\n",
       "13                           264\n",
       "MT                           213\n",
       "18                           209\n",
       "CHR_HSCHR6_MHC_DBB_CTG1      102\n",
       "CHR_HSCHR6_MHC_COX_CTG1       52\n",
       "CHR_HSCHR6_MHC_APD_CTG1       18\n",
       "CHR_HSCHR6_MHC_MANN_CTG1      16\n",
       "CHR_HSCHR6_MHC_MCF_CTG1       15\n",
       "CHR_HSCHR6_MHC_SSTO_CTG1      13\n",
       "CHR_HSCHR6_MHC_QBL_CTG1       11\n",
       "Y                              6\n",
       "Name: chromosome, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_coords = pd.read_csv(\"../../external/clash/clash_grch38_coordinates.bed\", sep=\"\\t\", header=None)\n",
    "new_coords.columns = [\"chromosome\", \"start\", \"end\"]\n",
    "\n",
    "new_coords.chromosome.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CLASH df original length: 18k\n",
    "\n",
    "len after transforming coordinates via ENSEMBL: 23k\n",
    "\n",
    "we shouldn\"t transform the data?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_ID</th>\n",
       "      <th>miRNA_start</th>\n",
       "      <th>miRNA_end</th>\n",
       "      <th>miRNA_seq</th>\n",
       "      <th>mRNA_start</th>\n",
       "      <th>mRNA_end_extended</th>\n",
       "      <th>mRNA_seq_extended</th>\n",
       "      <th>chimeras_decompressed</th>\n",
       "      <th>experiments</th>\n",
       "      <th>experiments_list</th>\n",
       "      <th>...</th>\n",
       "      <th>CLASH_cluster_ovlp</th>\n",
       "      <th>PAR_CLIP_cluster_ovlp</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>ensg</th>\n",
       "      <th>enst</th>\n",
       "      <th>strand</th>\n",
       "      <th>ensembl_sequence</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23289</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>2841486</td>\n",
       "      <td>2866956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23290</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>12904831</td>\n",
       "      <td>12910913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23291</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>12910913</td>\n",
       "      <td>12920478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23292</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>12904831</td>\n",
       "      <td>12910913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23293</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Y</td>\n",
       "      <td>12910913</td>\n",
       "      <td>12920478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      seq_ID  miRNA_start  miRNA_end miRNA_seq  mRNA_start  mRNA_end_extended  \\\n",
       "23289    NaN          NaN        NaN       NaN         NaN                NaN   \n",
       "23290    NaN          NaN        NaN       NaN         NaN                NaN   \n",
       "23291    NaN          NaN        NaN       NaN         NaN                NaN   \n",
       "23292    NaN          NaN        NaN       NaN         NaN                NaN   \n",
       "23293    NaN          NaN        NaN       NaN         NaN                NaN   \n",
       "\n",
       "      mRNA_seq_extended  chimeras_decompressed  experiments experiments_list  \\\n",
       "23289               NaN                    NaN          NaN              NaN   \n",
       "23290               NaN                    NaN          NaN              NaN   \n",
       "23291               NaN                    NaN          NaN              NaN   \n",
       "23292               NaN                    NaN          NaN              NaN   \n",
       "23293               NaN                    NaN          NaN              NaN   \n",
       "\n",
       "       ...  CLASH_cluster_ovlp  PAR_CLIP_cluster_ovlp mirna_name  ensg  enst  \\\n",
       "23289  ...                 NaN                    NaN        NaN   NaN   NaN   \n",
       "23290  ...                 NaN                    NaN        NaN   NaN   NaN   \n",
       "23291  ...                 NaN                    NaN        NaN   NaN   NaN   \n",
       "23292  ...                 NaN                    NaN        NaN   NaN   NaN   \n",
       "23293  ...                 NaN                    NaN        NaN   NaN   NaN   \n",
       "\n",
       "       strand  ensembl_sequence  chromosome     start       end  \n",
       "23289     NaN               NaN           Y   2841486   2866956  \n",
       "23290     NaN               NaN           Y  12904831  12910913  \n",
       "23291     NaN               NaN           Y  12910913  12920478  \n",
       "23292     NaN               NaN           Y  12904831  12910913  \n",
       "23293     NaN               NaN           Y  12910913  12920478  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[32m15:26:05\u001b[0m \u001b[1m|\u001b[0m \u001b[34mWARNING\u001b[0m \u001b[1m|\u001b[0m Could not find 'GRCh38.annotation.bed(.gz)' in directory /home/nazif/.local/share/genomes/GRCh38. Methods using this file won't work!\n",
      "\u001b[32m15:26:05\u001b[0m \u001b[1m|\u001b[0m \u001b[34mWARNING\u001b[0m \u001b[1m|\u001b[0m Could not find 'GRCh38.annotation.gtf(.gz)' in directory /home/nazif/.local/share/genomes/GRCh38. Methods using this file won't work!\n"
     ]
    }
   ],
   "source": [
    "import genomepy\n",
    "a = genomepy.Annotation(\"GRCh38\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
