# Topics

- TargetScan repos found at GitHub
  - Kathy Lin's TargetScan repo
  - Katyh Lin's miRNA_models repo
  - Vikram Agarwal's repo
- Packages to cover
  - scanmiR, an R package
  - [mirWalk](../miRNA%20Databases/mirWalk.md)
- New terms
  - oncomiR
  - metastamiR
- Found the paper I asked the week before
  - Downloaded the whole book and skimmed through it
  - [Machine Learning Based Methods and Best Practices of microRNA-Target Prediction and Validation](../Literature%20Notes/Machine%20Learning%20Based%20Methods%20and%20Best%20Practices%20of%20microRNA-Target%20Prediction%20and%20Validation.md)
  - [Turning Data to Knowledge, Online Tools, Databases, and Resources in microRNA Research](../Literature%20Notes/Turning%20Data%20to%20Knowledge,%20Online%20Tools,%20Databases,%20and%20Resources%20in%20microRNA%20Research.md)
- Another prediction database
  - [miRTar2GO](http://www.mirtar2go.org/downloadPage.html)
- miRNA-miRNA interactions
  - Seen in different cancer types
  - Chapter 9 of the pdf [(Advances in Experimental Medicine and Biology, 1385) Ulf Schmitz, Olaf Wolkenhauer, Julio Vera-González - Systems Biology of MicroRNAs in Cancer-Springer (2022)](../Literature%20Notes/(Advances%20in%20Experimental%20Medicine%20and%20Biology,%201385)%20Ulf%20Schmitz,%20Olaf%20Wolkenhauer,%20Julio%20Vera-González%20-%20Systems%20Biology%20of%20MicroRNAs%20in%20Cancer-Springer%20(2022).pdf)

# TargetScan Repos

## Kathy Lin's Repo

<https://github.com/kslin/targetscan>
Student of Prof. David Bartel, worked on last TargetScan paper

>*Papers of Bartel:*
>
>*(MicroRNAs: Target recognition and regulatory functions, 2009, <https://doi.org/10.1016/j.cell.2009.01.002>)*
>*(Predicting effective microRNA target sites in mammalian mRNAs, 2015, <https://doi.org/10.7554/eLife.05005>)*
>
>*(The biochemical basis of microRNA targeting efficacy, 2019, <https://doi.org/10.1126/science.aav1741>)*

### This Repo Has Following Folders

- PCT parameters
  - Has PCT (the probability of conserved targeting as described in [Friedman et al., 2009](http://genome.cshlp.org/content/19/1/92.long)), parameters for all possible seeds

- **calculate_bins** have the code for calculating bins

- **calculate_features** has the code for calculating features
  - TA_SPS_by_seed_region.txt has seed pairing stability (SPS) score for every possible kmers & their target abundance (TA) scores
  - feature_helpers.py has helper functions like getting energy from ViennaRNA package, 3' pairing score, local AU score
- **calculate_predictions** makes predictions

### Kathy Lin's miRNA_models Repo

This repo contains the code of their 2019 paper, with CNN

## Vikram Agarwal's Repo

TargetScan tools
May not be useful for us
<https://github.com/vagarwal87/TargetScanTools>

# Packages

## scanMiR

*(scanMiR: A biochemically based toolkit for versatile and efficient microRNA target prediction, 2022, <https://doi.org/10.1093/bioinformatics/btac110>)*

scanMiR is an R package for predicting miRNA targets. Online tool located [here](https://ethz-ins.org/scanMiR/). Bioconductor link [here](https://www.bioconductor.org/packages/devel/bioc/vignettes/scanMiR/inst/doc/scanning.html).

scanMiR uses the biochemical model of [Mcgeary & Lin 2019](https://www.science.org/doi/10.1126/science.aav1741).

## mirWalk

*(miRWalk: An online resource for prediction of microRNA binding sites, 2018, <https://doi.org/10.1371/journal.pone.0206239>)*

Detailed info about [mirWalk](../miRNA%20Databases/mirWalk.md) found here.

# Footnote

The meeting after I went to Konya
