{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing dependencies\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "def import_fasta(fasta):\n",
    "    \"\"\" imports fasta file into a python string\n",
    "\n",
    "    Args:\n",
    "        fasta (_type_): relative path of fasta file\n",
    "\n",
    "    Returns:\n",
    "        _type_: python string containing sequence\n",
    "    \"\"\"\n",
    "    with open(fasta) as f:\n",
    "        records = SeqIO.parse(f, \"fasta\")\n",
    "        transcripts = [str(rec.seq) for rec in records]\n",
    "        print(\"sequence loaded\")\n",
    "        return \"\".join(transcripts)\n",
    "\n",
    "\n",
    "def complement(string):\n",
    "\n",
    "    result = \"\"\n",
    "\n",
    "    for nuc in string:\n",
    "        if nuc == \"A\":\n",
    "            result += \"U\"\n",
    "        elif nuc == \"U\":\n",
    "            result += \"A\"\n",
    "        elif nuc == \"C\":\n",
    "            result += \"G\"\n",
    "        elif nuc == \"G\":\n",
    "            result += \"C\"\n",
    "\n",
    "    return result\n",
    "\n",
    "\n",
    "def uracil_to_thymine(string):\n",
    "\n",
    "    return \"\".join(\"T\" if nuc == \"U\" else nuc for nuc in string)\n",
    "\n",
    "\n",
    "\n",
    "def generate_kmers(sequence, k):\n",
    "    for nucleotide in range(len(sequence) - k + 1):\n",
    "        yield sequence[nucleotide: nucleotide + k]\n",
    "\n",
    "\n",
    "def generate_seed_kmer_dictionary(k, mirna_df):\n",
    "    target_seed = []\n",
    "    target_name = []\n",
    "\n",
    "    if k == 6:\n",
    "        target_name.extend(mirna_df[\"name\"].tolist())\n",
    "        target_seed.extend(mirna_df[\"6mer\"].tolist())\n",
    "\n",
    "        target_name.extend(mirna_df[\"name\"].tolist())\n",
    "        target_seed.extend(mirna_df[\"offset_6mer\"].tolist())\n",
    "\n",
    "    if k == 7:\n",
    "        target_name.extend(mirna_df[\"name\"].tolist())\n",
    "        target_seed.extend(mirna_df[\"7mer-A1\"].tolist())\n",
    "\n",
    "        target_name.extend(mirna_df[\"name\"].tolist())\n",
    "        target_seed.extend(mirna_df[\"7mer-m8\"].tolist())\n",
    "\n",
    "    if k == 8:\n",
    "        target_name.extend(mirna_df[\"name\"].tolist())\n",
    "        target_seed.extend(mirna_df[\"8mer\"].tolist())\n",
    "\n",
    "        # dictionary comprehension\n",
    "    return {target_seed[i]: target_name[i] for i in range(len(target_name))}\n",
    "\n",
    "\n",
    "def find_MRE_positions(kmer_generator, kmer_dictionary, k):\n",
    "    kmer_list = []\n",
    "    mirna_name = []\n",
    "    starting_position = []\n",
    "    ending_position = []\n",
    "\n",
    "    for count, kmer in enumerate(kmer_generator, start=0):\n",
    "\n",
    "        if kmer in kmer_dictionary:\n",
    "\n",
    "            position_of_first_nucleotide = count + 1\n",
    "            position_of_last_nucleotide = count + k\n",
    "\n",
    "            kmer_list.append(kmer)\n",
    "            mirna_name.append(kmer_dictionary[kmer])\n",
    "            starting_position.append(position_of_first_nucleotide)\n",
    "            ending_position.append(position_of_last_nucleotide)\n",
    "\n",
    "    return pd.DataFrame(list(zip(kmer_list, mirna_name, starting_position, ending_position)), columns=[\"kmer\", \"name\", \"start\", \"end\"])\n",
    "\n",
    "\n",
    "def run_script(k, fasta, mirna_fasta):\n",
    "\n",
    "    # DNA branch\n",
    "    sequence = import_fasta(fasta)\n",
    "    kmer_generator = generate_kmers(sequence, k)\n",
    "\n",
    "    # miRNA branch\n",
    "    mirna_df = create_mirna_database(mirna_fasta)\n",
    "    kmer_dictionary = generate_seed_kmer_dictionary(k, mirna_df)\n",
    "\n",
    "    return find_MRE_positions(kmer_generator, kmer_dictionary, k), mirna_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_mirna_database(fasta):\n",
    "    with open(fasta) as f:\n",
    "\n",
    "        # initializing lists to zip into a df\n",
    "        names = []\n",
    "        raw_3to5 = []\n",
    "        mRNA_sequences = []\n",
    "        cDNA_sequences = []\n",
    "\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "\n",
    "            name = str(record.id)\n",
    "            if name.startswith(\"hsa\"):  # drops all non-human entries\n",
    "                sequence = str(record.seq)\n",
    "\n",
    "                # adding name\n",
    "                names.append(name)\n",
    "\n",
    "                # adding raw 3 to 5 sequence\n",
    "                raw_3to5.append(sequence[::-1])\n",
    "\n",
    "                # adding mRNA sequence\n",
    "                mRNA_sequences.append(complement(sequence[::-1]))\n",
    "\n",
    "                # adding cDNA sequence\n",
    "                cDNA_sequences.append(uracil_to_thymine(sequence[::-1]))\n",
    "\n",
    "    # zipping lists into a pandas df\n",
    "    colnames = [\"name\", \"3to5_miRNA_seq\", \"mRNA_seq\", \"cDNA_seq\"]\n",
    "    df = pd.DataFrame(\n",
    "        list(zip(names, raw_3to5, mRNA_sequences, cDNA_sequences)), columns=colnames)\n",
    "\n",
    "    # deriving MRE sequences and adding as a new column\n",
    "    \"\"\" df[\"MRE\"] = df[\"cDNA_seq\"].str[-9:-1] \"\"\"\n",
    "\n",
    "    # deriving kmer columns\n",
    "    df[\"8mer\"] = df[\"cDNA_seq\"].str[-8:]\n",
    "    df[\"7mer-A1\"] = df[\"cDNA_seq\"].str[-7:]\n",
    "    df[\"7mer-m8\"] = df[\"cDNA_seq\"].str[-8:-1]\n",
    "    df[\"6mer\"] = df[\"cDNA_seq\"].str[-7:-1]\n",
    "    df[\"offset_6mer\"] = df[\"cDNA_seq\"].str[-8:-2]\n",
    "\n",
    "    # dropping 8mer and 7mer-A1 values that doesn't start with an A\n",
    "    #df[\"8mer\"].replace({\"[C,T,G]$\": None}, regex=True, inplace=True)\n",
    "    #df[\"7mer-A1\"].replace({\"[C,T,G]$\": None}, regex=True, inplace=True)\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def alt_create_mirna_database(fasta):\n",
    "    with open(fasta) as f:\n",
    "\n",
    "        # initializing lists to zip into a df\n",
    "        names = []\n",
    "        raw_3to5 = []\n",
    "\n",
    "        \n",
    "\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "\n",
    "            name = str(record.id)\n",
    "            if name.startswith(\"hsa\"):  # drops all non-human entries\n",
    "                sequence = str(record.seq)\n",
    "\n",
    "                # adding name\n",
    "                names.append(name)\n",
    "\n",
    "                # adding raw 3 to 5 sequence\n",
    "                raw_3to5.append(sequence[::-1])\n",
    "    \n",
    "    return names, raw_3to5\n",
    "\n",
    "def alt_run_script(k, fasta, mirna_fasta):\n",
    "\n",
    "    # DNA branch\n",
    "    sequence = import_fasta(fasta)\n",
    "    kmer_generator = generate_kmers(sequence, k)\n",
    "\n",
    "    # miRNA branch\n",
    "    mirna_df = create_mirna_database(mirna_fasta)\n",
    "    kmer_dictionary = generate_seed_kmer_dictionary(k, mirna_df)\n",
    "\n",
    "    return find_MRE_positions(kmer_generator, kmer_dictionary, k), mirna_df\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Files help:\n",
    "\n",
    "Fasta of GRCh38, chromosome 1, soft masked downloaded from:\n",
    "\n",
    "https://ftp.ensembl.org/pub/current_fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna_sm.chromosome.1.fa.gz\n",
    "\n",
    "Fasta of mature miRNA sequences downloaded from:\n",
    "\n",
    "https://www.mirbase.org/ftp.shtml\n",
    "\n",
    "## How to run the script:\n",
    "\n",
    "1. enter k size\n",
    "2. point to the target fasta\n",
    "3. point to the miRNA database fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "k = 7\n",
    "fasta = \"../sequences/grch38/chromosomes/Homo_sapiens.GRCh38.dna_sm.chromosome.1.fa\"\n",
    "hmga2 = \"../sequences/HMGA2_ENST00000403681_7_sequence.fa\"\n",
    "mirna_fasta = \"../sequences/mirna/mature_miRNAs.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence loaded\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>3to5_miRNA_seq</th>\n",
       "      <th>mRNA_seq</th>\n",
       "      <th>cDNA_seq</th>\n",
       "      <th>8mer</th>\n",
       "      <th>7mer-A1</th>\n",
       "      <th>7mer-m8</th>\n",
       "      <th>6mer</th>\n",
       "      <th>offset_6mer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>UUGAUAUGUUGGAUGAUGGAGU</td>\n",
       "      <td>AACUAUACAACCUACUACCUCA</td>\n",
       "      <td>TTGATATGTTGGATGATGGAGT</td>\n",
       "      <td>GATGGAGT</td>\n",
       "      <td>ATGGAGT</td>\n",
       "      <td>GATGGAG</td>\n",
       "      <td>ATGGAG</td>\n",
       "      <td>GATGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-3p</td>\n",
       "      <td>CUUUCUGUCAUCUAACAUAUC</td>\n",
       "      <td>GAAAGACAGUAGAUUGUAUAG</td>\n",
       "      <td>CTTTCTGTCATCTAACATATC</td>\n",
       "      <td>AACATATC</td>\n",
       "      <td>ACATATC</td>\n",
       "      <td>AACATAT</td>\n",
       "      <td>ACATAT</td>\n",
       "      <td>AACATA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-2-3p</td>\n",
       "      <td>CCUUUCGAUCCUCCGACAUGUC</td>\n",
       "      <td>GGAAAGCUAGGAGGCUGUACAG</td>\n",
       "      <td>CCTTTCGATCCTCCGACATGTC</td>\n",
       "      <td>GACATGTC</td>\n",
       "      <td>ACATGTC</td>\n",
       "      <td>GACATGT</td>\n",
       "      <td>ACATGT</td>\n",
       "      <td>GACATG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>UUGGUGUGUUGGAUGAUGGAGU</td>\n",
       "      <td>AACCACACAACCUACUACCUCA</td>\n",
       "      <td>TTGGTGTGTTGGATGATGGAGT</td>\n",
       "      <td>GATGGAGT</td>\n",
       "      <td>ATGGAGT</td>\n",
       "      <td>GATGGAG</td>\n",
       "      <td>ATGGAG</td>\n",
       "      <td>GATGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-3p</td>\n",
       "      <td>CCCUUCCGUCAUCCAACAUAUC</td>\n",
       "      <td>GGGAAGGCAGUAGGUUGUAUAG</td>\n",
       "      <td>CCCTTCCGTCATCCAACATATC</td>\n",
       "      <td>AACATATC</td>\n",
       "      <td>ACATATC</td>\n",
       "      <td>AACATAT</td>\n",
       "      <td>ACATAT</td>\n",
       "      <td>AACATA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              name          3to5_miRNA_seq                mRNA_seq  \\\n",
       "0    hsa-let-7a-5p  UUGAUAUGUUGGAUGAUGGAGU  AACUAUACAACCUACUACCUCA   \n",
       "1    hsa-let-7a-3p   CUUUCUGUCAUCUAACAUAUC   GAAAGACAGUAGAUUGUAUAG   \n",
       "2  hsa-let-7a-2-3p  CCUUUCGAUCCUCCGACAUGUC  GGAAAGCUAGGAGGCUGUACAG   \n",
       "3    hsa-let-7b-5p  UUGGUGUGUUGGAUGAUGGAGU  AACCACACAACCUACUACCUCA   \n",
       "4    hsa-let-7b-3p  CCCUUCCGUCAUCCAACAUAUC  GGGAAGGCAGUAGGUUGUAUAG   \n",
       "\n",
       "                 cDNA_seq      8mer  7mer-A1  7mer-m8    6mer offset_6mer  \n",
       "0  TTGATATGTTGGATGATGGAGT  GATGGAGT  ATGGAGT  GATGGAG  ATGGAG      GATGGA  \n",
       "1   CTTTCTGTCATCTAACATATC  AACATATC  ACATATC  AACATAT  ACATAT      AACATA  \n",
       "2  CCTTTCGATCCTCCGACATGTC  GACATGTC  ACATGTC  GACATGT  ACATGT      GACATG  \n",
       "3  TTGGTGTGTTGGATGATGGAGT  GATGGAGT  ATGGAGT  GATGGAG  ATGGAG      GATGGA  \n",
       "4  CCCTTCCGTCATCCAACATATC  AACATATC  ACATATC  AACATAT  ACATAT      AACATA  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results, mirna_df = run_script(k, hmga2, mirna_fasta)\n",
    "\n",
    "\"\"\" results.to_csv(\"file_name\", sep='\\t')\n",
    "mirna_df.to_csv(\"mirna_df.csv\", sep='\\t')\n",
    " \"\"\"\n",
    "\n",
    "mirna_df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[['hsa-miR-4458', 'AAGAAGGUGUGGAUGGAGA', 'UUCUUCCACACCUACCUCU', 'AAGAAGGTGTGGATGGAGA', 'GATGGAGA', 'ATGGAGA', 'GATGGAG', 'ATGGAG', 'GATGGA']]\n",
      "[['hsa-miR-4500', 'UUCUUUGAUGAUGGAGU', 'AAGAAACUACUACCUCA', 'TTCTTTGATGATGGAGT', 'GATGGAGT', 'ATGGAGT', 'GATGGAG', 'ATGGAG', 'GATGGA']]\n"
     ]
    }
   ],
   "source": [
    "mir_4458 = mirna_df[mirna_df[\"name\"] == \"hsa-miR-4458\"].values.tolist()\n",
    "mir_4500 = mirna_df[mirna_df[\"name\"] == \"hsa-miR-4500\"].values.tolist()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print(mir_4458)\n",
    "print(mir_4500)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "0a7314db4cd9cbf8825ff67208d8d351ac5340431b699125bf50d4fb6663edf3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
