{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from fuc import pybed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster.ID</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>N</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene.id</th>\n",
       "      <th>gene.symbol</th>\n",
       "      <th>region</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>...</th>\n",
       "      <th>target.map</th>\n",
       "      <th>MFE</th>\n",
       "      <th>duplex.start</th>\n",
       "      <th>seed match</th>\n",
       "      <th>seed position</th>\n",
       "      <th>bulged miRNA position</th>\n",
       "      <th>bulged target position</th>\n",
       "      <th>mismatch seed position</th>\n",
       "      <th>mismatched or bulged nucleotide(s)</th>\n",
       "      <th>k.group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_100127966_100128022_hsa-miR-16-5p</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100127966</td>\n",
       "      <td>100128041</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>intron</td>\n",
       "      <td>hsa-miR-16-5p</td>\n",
       "      <td>...</td>\n",
       "      <td>|||||||----W||||W|||-</td>\n",
       "      <td>-19.5</td>\n",
       "      <td>6</td>\n",
       "      <td>7merm8</td>\n",
       "      <td>20.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_100154790_100154826_hsa-miR-92a-3p</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100154790</td>\n",
       "      <td>100154865</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hsa-miR-92a-3p</td>\n",
       "      <td>...</td>\n",
       "      <td>-|||||---|||-||||-||WW|-</td>\n",
       "      <td>-27.2</td>\n",
       "      <td>8</td>\n",
       "      <td>7merA1.indel</td>\n",
       "      <td>26.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>5B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_100160176_100160232_hsa-miR-30a-5p</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100160176</td>\n",
       "      <td>100160251</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>hsa-miR-30a-5p</td>\n",
       "      <td>...</td>\n",
       "      <td>-||||----|||-</td>\n",
       "      <td>-9.4</td>\n",
       "      <td>54</td>\n",
       "      <td>6merA1.mm</td>\n",
       "      <td>32.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>A,A</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_100260813_100260845_hsa-miR-500a-5p</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100260813</td>\n",
       "      <td>100260888</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>23443.0</td>\n",
       "      <td>SLC35A3</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>hsa-miR-500a-5p</td>\n",
       "      <td>...</td>\n",
       "      <td>||||||----|||||W-</td>\n",
       "      <td>-22.0</td>\n",
       "      <td>6</td>\n",
       "      <td>7merA1</td>\n",
       "      <td>17.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_100262727_100262757_hsa-miR-181a-5p</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100262727</td>\n",
       "      <td>100262802</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>23443.0</td>\n",
       "      <td>SLC35A3</td>\n",
       "      <td>3'UTR||intron</td>\n",
       "      <td>hsa-miR-181a-5p</td>\n",
       "      <td>...</td>\n",
       "      <td>-||||||-W||||-||||-</td>\n",
       "      <td>-23.6</td>\n",
       "      <td>4</td>\n",
       "      <td>6mer_off</td>\n",
       "      <td>16.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 cluster.ID   chr      start        end  N  \\\n",
       "0    chr1_100127966_100128022_hsa-miR-16-5p  chr1  100127966  100128041  1   \n",
       "1   chr1_100154790_100154826_hsa-miR-92a-3p  chr1  100154790  100154865  1   \n",
       "2   chr1_100160176_100160232_hsa-miR-30a-5p  chr1  100160176  100160251  1   \n",
       "3  chr1_100260813_100260845_hsa-miR-500a-5p  chr1  100260813  100260888  1   \n",
       "4  chr1_100262727_100262757_hsa-miR-181a-5p  chr1  100262727  100262802  1   \n",
       "\n",
       "  strand  gene.id gene.symbol         region            miRNA  ...  \\\n",
       "0      +    178.0         AGL         intron    hsa-miR-16-5p  ...   \n",
       "1      +    178.0         AGL            CDS   hsa-miR-92a-3p  ...   \n",
       "2      +    178.0         AGL          3'UTR   hsa-miR-30a-5p  ...   \n",
       "3      +  23443.0     SLC35A3          3'UTR  hsa-miR-500a-5p  ...   \n",
       "4      +  23443.0     SLC35A3  3'UTR||intron  hsa-miR-181a-5p  ...   \n",
       "\n",
       "                 target.map   MFE  duplex.start    seed match seed position  \\\n",
       "0     |||||||----W||||W|||- -19.5             6        7merm8          20.0   \n",
       "1  -|||||---|||-||||-||WW|- -27.2             8  7merA1.indel          26.0   \n",
       "2             -||||----|||-  -9.4            54     6merA1.mm          32.0   \n",
       "3         ||||||----|||||W- -22.0             6        7merA1          17.0   \n",
       "4       -||||||-W||||-||||- -23.6             4      6mer_off          16.0   \n",
       "\n",
       "   bulged miRNA position  bulged target position mismatch seed position  \\\n",
       "0                    NaN                     NaN                    NaN   \n",
       "1                    2.0                     NaN                    NaN   \n",
       "2                    NaN                     NaN                    4.0   \n",
       "3                    NaN                     NaN                    NaN   \n",
       "4                    NaN                     NaN                    NaN   \n",
       "\n",
       "   mismatched or bulged nucleotide(s) k.group  \n",
       "0                                 NaN       2  \n",
       "1                                   A      5B  \n",
       "2                                 A,A       4  \n",
       "3                                 NaN       2  \n",
       "4                                 NaN       2  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"clear_data.csv\")\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### rename columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rename_dict = {\n",
    "    'cluster.ID': 'cluster_id',\n",
    "    'chr': 'chromosome',\n",
    "    'start': 'start',\n",
    "    'end': 'end',\n",
    "    'N': 'N_value',\n",
    "    'strand': 'strand_direction',\n",
    "    'gene.id': 'gene_id',\n",
    "    'gene.symbol': 'gene_symbol',\n",
    "    'region': 'region_type',\n",
    "    'miRNA': 'miRNA_sequence',\n",
    "    'miR.map': 'miRNA_map',\n",
    "    'target.map': 'target_map',\n",
    "    'MFE': 'minimum_free_energy',\n",
    "    'duplex.start': 'duplex_start_position',\n",
    "    'seed match': 'seed_match',\n",
    "    'seed position': 'seed_position',\n",
    "    'bulged miRNA position': 'bulged_miRNA_position',\n",
    "    'bulged target position': 'bulged_target_position',\n",
    "    'mismatch seed position': 'mismatch_seed_position',\n",
    "    'mismatched or bulged nucleotide(s)': 'mismatch_bulged_nucleotides',\n",
    "    'k.group': 'k_group'\n",
    "}\n",
    "# rename columns\n",
    "df = df.rename(columns=rename_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# uncomment if you want to generate the bed file\n",
    "\n",
    "\n",
    "# # filter & rename columns for pybed\n",
    "# coord_df = df[[\"chromosome\", \"start\", \"end\"]]\n",
    "# coord_df.columns = ['Chromosome', 'Start', 'End']\n",
    "\n",
    "# # call pybed to turn into .bed file\n",
    "# bf = pybed.BedFrame.from_frame(meta=[], data=coord_df)\n",
    "# bf.to_file('clear_hg18_coordinates.bed')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### replacing old coordinates with new ones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop old coords\n",
    "df.drop(['chromosome', 'start', 'end'], axis=1, inplace=True)\n",
    "\n",
    "\n",
    "# read & merge new ones\n",
    "new_coords = pd.read_csv(\"../../external/clear/clear_grch38_coordinates.bed\", sep=\"\\t\", header=None)\n",
    "new_coords.columns = ['chromosome', 'start', 'end']\n",
    "\n",
    "df = pd.concat([df, new_coords], axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32712\n",
      "14\n"
     ]
    }
   ],
   "source": [
    "print(len(df))\n",
    "print(df.start.isna().sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32698\n"
     ]
    }
   ],
   "source": [
    "# drop rows with NaN values in the \"start\" column. These coordinates can't be found by ENSEMBL's coordinate transformer\n",
    "df = df.dropna(subset=['start'])\n",
    "\n",
    "# convert the cols to integer as it don't have any NaN left\n",
    "df[\"start\"] = df[\"start\"].astype(int)\n",
    "df[\"end\"] = df[\"end\"].astype(int)\n",
    "\n",
    "print(len(df))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### converting mRNA and miRNA maps into alignment strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['cluster_id', 'N_value', 'strand_direction', 'gene_id', 'gene_symbol',\n",
       "       'region_type', 'miRNA_sequence', 'miRNA_map', 'target_map',\n",
       "       'minimum_free_energy', 'duplex_start_position', 'seed_match',\n",
       "       'seed_position', 'bulged_miRNA_position', 'bulged_target_position',\n",
       "       'mismatch_seed_position', 'mismatch_bulged_nucleotides', 'k_group',\n",
       "       'chromosome', 'start', 'end'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>N_value</th>\n",
       "      <th>strand_direction</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>region_type</th>\n",
       "      <th>miRNA_sequence</th>\n",
       "      <th>minimum_free_energy</th>\n",
       "      <th>duplex_start_position</th>\n",
       "      <th>seed_match</th>\n",
       "      <th>...</th>\n",
       "      <th>bulged_miRNA_position</th>\n",
       "      <th>bulged_target_position</th>\n",
       "      <th>mismatch_seed_position</th>\n",
       "      <th>mismatch_bulged_nucleotides</th>\n",
       "      <th>k_group</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>alignment_string</th>\n",
       "      <th>mirna_alignment_string</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_100127966_100128022_hsa-miR-16-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>intron</td>\n",
       "      <td>hsa-miR-16-5p</td>\n",
       "      <td>-19.5</td>\n",
       "      <td>6</td>\n",
       "      <td>7merm8</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>99889822</td>\n",
       "      <td>99889897</td>\n",
       "      <td>111111100002111121110</td>\n",
       "      <td>111111102111121001100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_100154790_100154826_hsa-miR-92a-3p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hsa-miR-92a-3p</td>\n",
       "      <td>-27.2</td>\n",
       "      <td>8</td>\n",
       "      <td>7merA1.indel</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>5B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>99916646</td>\n",
       "      <td>99916721</td>\n",
       "      <td>011111000111011110112210</td>\n",
       "      <td>011111000111111111221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_100160176_100160232_hsa-miR-30a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>hsa-miR-30a-5p</td>\n",
       "      <td>-9.4</td>\n",
       "      <td>54</td>\n",
       "      <td>6merA1.mm</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>A,A</td>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>99922032</td>\n",
       "      <td>99922107</td>\n",
       "      <td>0111100001110</td>\n",
       "      <td>000000000000011110111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_100260813_100260845_hsa-miR-500a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>23443.0</td>\n",
       "      <td>SLC35A3</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>hsa-miR-500a-5p</td>\n",
       "      <td>-22.0</td>\n",
       "      <td>6</td>\n",
       "      <td>7merA1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100022669</td>\n",
       "      <td>100022744</td>\n",
       "      <td>11111100001111120</td>\n",
       "      <td>1111110000011111200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_100262727_100262757_hsa-miR-181a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>23443.0</td>\n",
       "      <td>SLC35A3</td>\n",
       "      <td>3'UTR||intron</td>\n",
       "      <td>hsa-miR-181a-5p</td>\n",
       "      <td>-23.6</td>\n",
       "      <td>4</td>\n",
       "      <td>6mer_off</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100024583</td>\n",
       "      <td>100024658</td>\n",
       "      <td>0111111021111011110</td>\n",
       "      <td>0111111002111101111000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32693</th>\n",
       "      <td>chrY_14801934_14801989_hsa-miR-17-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deep_intergenic</td>\n",
       "      <td>hsa-miR-17-5p</td>\n",
       "      <td>-9.6</td>\n",
       "      <td>52</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2925180</td>\n",
       "      <td>2925255</td>\n",
       "      <td>0111011001210001110110</td>\n",
       "      <td>0001110110012100111110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32694</th>\n",
       "      <td>chrY_15577527_15577583_hsa-miR-196a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deep_intergenic</td>\n",
       "      <td>hsa-miR-196a-5p</td>\n",
       "      <td>-17.7</td>\n",
       "      <td>28</td>\n",
       "      <td>7merA1.indel</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>6</td>\n",
       "      <td>chrY</td>\n",
       "      <td>6763687</td>\n",
       "      <td>6763762</td>\n",
       "      <td>011111001110001110000000021110</td>\n",
       "      <td>011111011101110021110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32695</th>\n",
       "      <td>chrY_16665792_16665848_hsa-miR-146a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deep_intergenic</td>\n",
       "      <td>hsa-miR-146a-5p</td>\n",
       "      <td>-17.3</td>\n",
       "      <td>9</td>\n",
       "      <td>6mer.indel</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>5B</td>\n",
       "      <td>chrY</td>\n",
       "      <td>7639977</td>\n",
       "      <td>7640052</td>\n",
       "      <td>111000110001111121110</td>\n",
       "      <td>111001100001110112111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32696</th>\n",
       "      <td>chrY_17760045_17760101_hsa-miR-194-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deep_intergenic</td>\n",
       "      <td>hsa-miR-194-5p</td>\n",
       "      <td>-15.7</td>\n",
       "      <td>43</td>\n",
       "      <td>6mer.indel</td>\n",
       "      <td>...</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>T</td>\n",
       "      <td>4</td>\n",
       "      <td>chrY</td>\n",
       "      <td>8584118</td>\n",
       "      <td>8584193</td>\n",
       "      <td>0111111000000101110110</td>\n",
       "      <td>000011111100111101100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32697</th>\n",
       "      <td>chrY_17943600_17943634_hsa-miR-34a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>deep_intergenic</td>\n",
       "      <td>hsa-miR-34a-5p</td>\n",
       "      <td>-18.1</td>\n",
       "      <td>43</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>8746321</td>\n",
       "      <td>8746396</td>\n",
       "      <td>121000011021111120</td>\n",
       "      <td>121000011021111120000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32698 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     cluster_id  N_value strand_direction  \\\n",
       "0        chr1_100127966_100128022_hsa-miR-16-5p        1                +   \n",
       "1       chr1_100154790_100154826_hsa-miR-92a-3p        1                +   \n",
       "2       chr1_100160176_100160232_hsa-miR-30a-5p        1                +   \n",
       "3      chr1_100260813_100260845_hsa-miR-500a-5p        1                +   \n",
       "4      chr1_100262727_100262757_hsa-miR-181a-5p        1                +   \n",
       "...                                         ...      ...              ...   \n",
       "32693      chrY_14801934_14801989_hsa-miR-17-5p        1                -   \n",
       "32694    chrY_15577527_15577583_hsa-miR-196a-5p        1                +   \n",
       "32695    chrY_16665792_16665848_hsa-miR-146a-5p        1                -   \n",
       "32696     chrY_17760045_17760101_hsa-miR-194-5p        1                -   \n",
       "32697     chrY_17943600_17943634_hsa-miR-34a-5p        1                +   \n",
       "\n",
       "       gene_id gene_symbol      region_type   miRNA_sequence  \\\n",
       "0        178.0         AGL           intron    hsa-miR-16-5p   \n",
       "1        178.0         AGL              CDS   hsa-miR-92a-3p   \n",
       "2        178.0         AGL            3'UTR   hsa-miR-30a-5p   \n",
       "3      23443.0     SLC35A3            3'UTR  hsa-miR-500a-5p   \n",
       "4      23443.0     SLC35A3    3'UTR||intron  hsa-miR-181a-5p   \n",
       "...        ...         ...              ...              ...   \n",
       "32693      NaN         NaN  deep_intergenic    hsa-miR-17-5p   \n",
       "32694      NaN         NaN  deep_intergenic  hsa-miR-196a-5p   \n",
       "32695      NaN         NaN  deep_intergenic  hsa-miR-146a-5p   \n",
       "32696      NaN         NaN  deep_intergenic   hsa-miR-194-5p   \n",
       "32697      NaN         NaN  deep_intergenic   hsa-miR-34a-5p   \n",
       "\n",
       "       minimum_free_energy  duplex_start_position    seed_match  ...  \\\n",
       "0                    -19.5                      6        7merm8  ...   \n",
       "1                    -27.2                      8  7merA1.indel  ...   \n",
       "2                     -9.4                     54     6merA1.mm  ...   \n",
       "3                    -22.0                      6        7merA1  ...   \n",
       "4                    -23.6                      4      6mer_off  ...   \n",
       "...                    ...                    ...           ...  ...   \n",
       "32693                 -9.6                     52           NaN  ...   \n",
       "32694                -17.7                     28  7merA1.indel  ...   \n",
       "32695                -17.3                      9    6mer.indel  ...   \n",
       "32696                -15.7                     43    6mer.indel  ...   \n",
       "32697                -18.1                     43           NaN  ...   \n",
       "\n",
       "       bulged_miRNA_position  bulged_target_position mismatch_seed_position  \\\n",
       "0                        NaN                     NaN                    NaN   \n",
       "1                        2.0                     NaN                    NaN   \n",
       "2                        NaN                     NaN                    4.0   \n",
       "3                        NaN                     NaN                    NaN   \n",
       "4                        NaN                     NaN                    NaN   \n",
       "...                      ...                     ...                    ...   \n",
       "32693                    NaN                     NaN                    NaN   \n",
       "32694                    2.0                     NaN                    NaN   \n",
       "32695                    3.0                     NaN                    NaN   \n",
       "32696                    3.0                     NaN                    NaN   \n",
       "32697                    NaN                     NaN                    NaN   \n",
       "\n",
       "       mismatch_bulged_nucleotides k_group chromosome      start        end  \\\n",
       "0                              NaN       2       chr1   99889822   99889897   \n",
       "1                                A      5B       chr1   99916646   99916721   \n",
       "2                              A,A       4       chr1   99922032   99922107   \n",
       "3                              NaN       2       chr1  100022669  100022744   \n",
       "4                              NaN       2       chr1  100024583  100024658   \n",
       "...                            ...     ...        ...        ...        ...   \n",
       "32693                          NaN       4       chrY    2925180    2925255   \n",
       "32694                            A       6       chrY    6763687    6763762   \n",
       "32695                            A      5B       chrY    7639977    7640052   \n",
       "32696                            T       4       chrY    8584118    8584193   \n",
       "32697                          NaN       2       chrY    8746321    8746396   \n",
       "\n",
       "                     alignment_string  mirna_alignment_string  \n",
       "0               111111100002111121110   111111102111121001100  \n",
       "1            011111000111011110112210   011111000111111111221  \n",
       "2                       0111100001110   000000000000011110111  \n",
       "3                   11111100001111120  1111110000011111200000  \n",
       "4                 0111111021111011110  0111111002111101111000  \n",
       "...                               ...                     ...  \n",
       "32693          0111011001210001110110  0001110110012100111110  \n",
       "32694  011111001110001110000000021110   011111011101110021110  \n",
       "32695           111000110001111121110   111001100001110112111  \n",
       "32696          0111111000000101110110   000011111100111101100  \n",
       "32697              121000011021111120   121000011021111120000  \n",
       "\n",
       "[32698 rows x 21 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>N_value</th>\n",
       "      <th>strand_direction</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>region_type</th>\n",
       "      <th>miRNA_sequence</th>\n",
       "      <th>minimum_free_energy</th>\n",
       "      <th>duplex_start_position</th>\n",
       "      <th>seed_match</th>\n",
       "      <th>...</th>\n",
       "      <th>bulged_miRNA_position</th>\n",
       "      <th>bulged_target_position</th>\n",
       "      <th>mismatch_seed_position</th>\n",
       "      <th>mismatch_bulged_nucleotides</th>\n",
       "      <th>k_group</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>alignment_string</th>\n",
       "      <th>mirna_alignment_string</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_100127966_100128022_hsa-miR-16-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>intron</td>\n",
       "      <td>hsa-miR-16-5p</td>\n",
       "      <td>-19.5</td>\n",
       "      <td>6</td>\n",
       "      <td>7merm8</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>99889822</td>\n",
       "      <td>99889897</td>\n",
       "      <td>111111100002111121110</td>\n",
       "      <td>111111102111121001100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_100154790_100154826_hsa-miR-92a-3p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>CDS</td>\n",
       "      <td>hsa-miR-92a-3p</td>\n",
       "      <td>-27.2</td>\n",
       "      <td>8</td>\n",
       "      <td>7merA1.indel</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A</td>\n",
       "      <td>5B</td>\n",
       "      <td>chr1</td>\n",
       "      <td>99916646</td>\n",
       "      <td>99916721</td>\n",
       "      <td>011111000111011110112210</td>\n",
       "      <td>011111000111111111221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_100160176_100160232_hsa-miR-30a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>178.0</td>\n",
       "      <td>AGL</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>hsa-miR-30a-5p</td>\n",
       "      <td>-9.4</td>\n",
       "      <td>54</td>\n",
       "      <td>6merA1.mm</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>A,A</td>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>99922032</td>\n",
       "      <td>99922107</td>\n",
       "      <td>0111100001110</td>\n",
       "      <td>000000000000011110111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_100260813_100260845_hsa-miR-500a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>23443.0</td>\n",
       "      <td>SLC35A3</td>\n",
       "      <td>3'UTR</td>\n",
       "      <td>hsa-miR-500a-5p</td>\n",
       "      <td>-22.0</td>\n",
       "      <td>6</td>\n",
       "      <td>7merA1</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100022669</td>\n",
       "      <td>100022744</td>\n",
       "      <td>11111100001111120</td>\n",
       "      <td>1111110000011111200000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_100262727_100262757_hsa-miR-181a-5p</td>\n",
       "      <td>1</td>\n",
       "      <td>+</td>\n",
       "      <td>23443.0</td>\n",
       "      <td>SLC35A3</td>\n",
       "      <td>3'UTR||intron</td>\n",
       "      <td>hsa-miR-181a-5p</td>\n",
       "      <td>-23.6</td>\n",
       "      <td>4</td>\n",
       "      <td>6mer_off</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100024583</td>\n",
       "      <td>100024658</td>\n",
       "      <td>0111111021111011110</td>\n",
       "      <td>0111111002111101111000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 cluster_id  N_value strand_direction  \\\n",
       "0    chr1_100127966_100128022_hsa-miR-16-5p        1                +   \n",
       "1   chr1_100154790_100154826_hsa-miR-92a-3p        1                +   \n",
       "2   chr1_100160176_100160232_hsa-miR-30a-5p        1                +   \n",
       "3  chr1_100260813_100260845_hsa-miR-500a-5p        1                +   \n",
       "4  chr1_100262727_100262757_hsa-miR-181a-5p        1                +   \n",
       "\n",
       "   gene_id gene_symbol    region_type   miRNA_sequence  minimum_free_energy  \\\n",
       "0    178.0         AGL         intron    hsa-miR-16-5p                -19.5   \n",
       "1    178.0         AGL            CDS   hsa-miR-92a-3p                -27.2   \n",
       "2    178.0         AGL          3'UTR   hsa-miR-30a-5p                 -9.4   \n",
       "3  23443.0     SLC35A3          3'UTR  hsa-miR-500a-5p                -22.0   \n",
       "4  23443.0     SLC35A3  3'UTR||intron  hsa-miR-181a-5p                -23.6   \n",
       "\n",
       "   duplex_start_position    seed_match  ...  bulged_miRNA_position  \\\n",
       "0                      6        7merm8  ...                    NaN   \n",
       "1                      8  7merA1.indel  ...                    2.0   \n",
       "2                     54     6merA1.mm  ...                    NaN   \n",
       "3                      6        7merA1  ...                    NaN   \n",
       "4                      4      6mer_off  ...                    NaN   \n",
       "\n",
       "   bulged_target_position mismatch_seed_position  mismatch_bulged_nucleotides  \\\n",
       "0                     NaN                    NaN                          NaN   \n",
       "1                     NaN                    NaN                            A   \n",
       "2                     NaN                    4.0                          A,A   \n",
       "3                     NaN                    NaN                          NaN   \n",
       "4                     NaN                    NaN                          NaN   \n",
       "\n",
       "  k_group chromosome      start        end          alignment_string  \\\n",
       "0       2       chr1   99889822   99889897     111111100002111121110   \n",
       "1      5B       chr1   99916646   99916721  011111000111011110112210   \n",
       "2       4       chr1   99922032   99922107             0111100001110   \n",
       "3       2       chr1  100022669  100022744         11111100001111120   \n",
       "4       2       chr1  100024583  100024658       0111111021111011110   \n",
       "\n",
       "   mirna_alignment_string  \n",
       "0   111111102111121001100  \n",
       "1   011111000111111111221  \n",
       "2   000000000000011110111  \n",
       "3  1111110000011111200000  \n",
       "4  0111111002111101111000  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_table = str.maketrans(\"-|W\", \"012\")\n",
    "\n",
    "df[\"alignment_string\"] = df[\"target_map\"].str.translate(trans_table)\n",
    "df[\"mirna_alignment_string\"] = df[\"miRNA_map\"].str.translate(trans_table)\n",
    "\n",
    "df = df.drop(columns=[\"target_map\", \"miRNA_map\"])\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"../../processed/clear/clear.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32698"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
