# Helwak 2013 CLASH

## Summary

MicroRNAs (miRNAs) play key roles in gene regulation, but reliable bioinformatic or experimental identification of their targets remains difficult. To provide an unbiased view of human miRNA targets, we developed a technique for ligation and sequencing of miRNA-target RNA duplexes associated with human AGO1. Here, we report data sets of more than 18,000 high-confidence miRNA-mRNA interactions. The binding of most miRNAs includes the 5′ seed region, but around 60% of seed interactions are noncanonical, containing bulged or mismatched nucleotides. Moreover, seed interactions are generally accompanied by specific, nonseed base pairing. 18% of miRNA-mRNA interactions involve the miRNA 3′ end, with little evidence for 5′ contacts, and some of these were functionally validated. Analyses of miRNA:mRNA base pairing showed that miRNA species systematically differ in their target RNA interactions, and strongly overrepresented motifs were found in the interaction sites of several miRNAs. We speculate that these affect the response of RISC to miRNA-target binding.

*(Mapping the human miRNA interactome by CLASH reveals frequent noncanonical binding, 2013, https://doi.org/10.1016/j.cell.2013.03.043)*

---
Agarwal 2015 also used this dataset.

*(Predicting effective microRNA target sites in mammalian mRNAs, 2015, https://doi.org/10.7554/eLife.05005)*

---
