{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing dependencies\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "\n",
    "\n",
    "def import_fasta(fasta):\n",
    "\n",
    "    with open(fasta) as f:\n",
    "        records = SeqIO.parse(f, \"fasta\")\n",
    "        transcripts = [dna_to_mrna(str(rec.seq)) for rec in records]\n",
    "        print(\"sequence loaded\")\n",
    "        return \"\".join(transcripts)\n",
    "\n",
    "\n",
    "def complement(string):\n",
    "\n",
    "    result = \"\"\n",
    "    for nuc in string:\n",
    "        if nuc == \"A\":\n",
    "            result += \"U\"\n",
    "        elif nuc == \"C\":\n",
    "            result += \"G\"\n",
    "        elif nuc == \"G\":\n",
    "            result += \"C\"\n",
    "        elif nuc in [\"U\", \"T\"]:\n",
    "            result += \"A\"\n",
    "    return result\n",
    "\n",
    "\n",
    "def uracil_to_thymine(string):\n",
    "    return \"\".join(\"T\" if nuc == \"U\" else nuc for nuc in string)\n",
    "\n",
    "\n",
    "def mirna_to_mrna(string):\n",
    "    return complement(string[::-1])\n",
    "\n",
    "\n",
    "def dna_to_mrna(string):\n",
    "    return \"\".join(\"U\" if nuc == \"T\" else nuc for nuc in string)\n",
    "\n",
    "\n",
    "def sliding_window(sequence, win_size):\n",
    "    for i in range(len(sequence) - win_size + 1):\n",
    "        yield sequence[i : i + win_size]\n",
    "\n",
    "\n",
    "def create_mirna_matrix(fasta):\n",
    "    with open(fasta) as f:\n",
    "\n",
    "        # initializing lists to zip into a df\n",
    "        names = []\n",
    "        sequences = []\n",
    "\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "\n",
    "            name = str(record.id)\n",
    "            if name.startswith(\"hsa\"):  # drops all non-human entries\n",
    "                sequence = str(record.seq)\n",
    "\n",
    "                # adding name\n",
    "                names.append(name)\n",
    "\n",
    "                # adding mRNA sequence\n",
    "                sequences.append(mirna_to_mrna(sequence))\n",
    "\n",
    "    return [names, sequences]\n",
    "\n",
    "\n",
    "def find_matches(sequence, mirna_matrix):  # sourcery skip: remove-redundant-if\n",
    "\n",
    "    # mirna sequences are already converted to reverse complement (mRNA)\n",
    "    mirna_names = mirna_matrix[0]\n",
    "    mirna_sequences = mirna_matrix[1]\n",
    "\n",
    "    mirna_seeds = [seq[-7:-1] for seq in mirna_sequences]\n",
    "\n",
    "    seed_dict = defaultdict(list)\n",
    "\n",
    "    for i in range(len(mirna_names)):\n",
    "        seed_dict[mirna_seeds[i]].append((mirna_names[i], mirna_sequences[i]))\n",
    "\n",
    "    # generates k=10 length chunks of mRNA\n",
    "    generator = sliding_window(dna_to_mrna(sequence), 10)\n",
    "\n",
    "    names = []\n",
    "    coordinates = []\n",
    "    seed_matches = []\n",
    "    chunks = []\n",
    "    match_types = []\n",
    "\n",
    "    # for loop\n",
    "    for c, chunk in enumerate(generator, start=1):\n",
    "\n",
    "        seed = chunk[2:8]\n",
    "\n",
    "        # if there's a seed match found\n",
    "        if seed in seed_dict:\n",
    "            for record in seed_dict[seed]:\n",
    "\n",
    "                mirna_sequence = seed_dict[seed][0][1]\n",
    "\n",
    "                if chunk[8] == \"A\" and chunk[1] == mirna_sequence[-8]:\n",
    "                    match_types.append(\"8mer\")\n",
    "                    coordinates.append(f\"{str(c + 1)}-{str(c + 8)}\")\n",
    "                    seed_matches.append(chunk[1:9])\n",
    "\n",
    "                elif chunk[8] == \"A\":\n",
    "                    match_types.append(\"7mer-A1\")\n",
    "                    coordinates.append(f\"{str(c + 2)}-{str(c + 8)}\")\n",
    "                    seed_matches.append(chunk[2:9])\n",
    "\n",
    "                elif chunk[1] == mirna_sequence[-8]:\n",
    "                    match_types.append(\"7mer-m8\")\n",
    "                    coordinates.append(f\"{str(c + 1)}-{str(c + 7)}\")\n",
    "                    seed_matches.append(chunk[1:8])\n",
    "\n",
    "                else:\n",
    "                    match_types.append(\"6mer\")\n",
    "                    coordinates.append(f\"{str(c + 2)}-{str(c + 7)}\")\n",
    "                    seed_matches.append(chunk[2:8])\n",
    "\n",
    "                chunks.append(chunk)\n",
    "                names.append(record[0])\n",
    "\n",
    "    return pd.DataFrame(\n",
    "        list(\n",
    "            zip(\n",
    "                names,\n",
    "                coordinates,\n",
    "                seed_matches,\n",
    "                chunks,\n",
    "                match_types,\n",
    "            )\n",
    "        ),\n",
    "        columns=[\n",
    "            \"name\",\n",
    "            \"coordinates\",\n",
    "            \"seed match\",\n",
    "            \"chunk\",\n",
    "            \"match_type\",\n",
    "        ],\n",
    "    )"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# running the script\n",
    "\n",
    "test case: 3' UTR of human HMGA2_ENST00000403681_7 transcript\n",
    "\n",
    "This gene is the example gene used in the TargetScan website, therefore I picked it.\n",
    "\n",
    "TargetScan doesn't have the results for **HMGA2_ENST00000403681_7**, instead, it has results for **HMGA2_ENST00000403681_2**. I couldn't access the version .2 of that transcript, so I went with .7 version. I don't know if these transcripts are exactly the same or have some differences.\n",
    "\n",
    "**question: How can I access that?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence loaded\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>seed match</th>\n",
       "      <th>chunk</th>\n",
       "      <th>match_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-miR-6789-3p</td>\n",
       "      <td>2-9</td>\n",
       "      <td>GGGCGCCA</td>\n",
       "      <td>GGGGCGCCAA</td>\n",
       "      <td>8mer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-miR-1306-3p</td>\n",
       "      <td>6-12</td>\n",
       "      <td>GCCAACG</td>\n",
       "      <td>CGCCAACGUU</td>\n",
       "      <td>7mer-m8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-miR-5680</td>\n",
       "      <td>17-23</td>\n",
       "      <td>AUUUCUA</td>\n",
       "      <td>CGAUUUCUAC</td>\n",
       "      <td>7mer-A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-miR-379-5p</td>\n",
       "      <td>20-25</td>\n",
       "      <td>UCUACC</td>\n",
       "      <td>UUUCUACCUC</td>\n",
       "      <td>6mer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-miR-4451</td>\n",
       "      <td>20-25</td>\n",
       "      <td>UCUACC</td>\n",
       "      <td>UUUCUACCUC</td>\n",
       "      <td>6mer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2302</th>\n",
       "      <td>hsa-miR-556-3p</td>\n",
       "      <td>2976-2982</td>\n",
       "      <td>GUAAUAA</td>\n",
       "      <td>GAGUAAUAAA</td>\n",
       "      <td>7mer-A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2303</th>\n",
       "      <td>hsa-miR-19a-5p</td>\n",
       "      <td>2991-2997</td>\n",
       "      <td>CAAAACA</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2304</th>\n",
       "      <td>hsa-miR-19b-1-5p</td>\n",
       "      <td>2991-2997</td>\n",
       "      <td>CAAAACA</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2305</th>\n",
       "      <td>hsa-miR-19b-2-5p</td>\n",
       "      <td>2991-2997</td>\n",
       "      <td>CAAAACA</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>hsa-miR-2052</td>\n",
       "      <td>2991-2997</td>\n",
       "      <td>CAAAACA</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2307 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  name coordinates seed match       chunk match_type\n",
       "0      hsa-miR-6789-3p         2-9   GGGCGCCA  GGGGCGCCAA       8mer\n",
       "1      hsa-miR-1306-3p        6-12    GCCAACG  CGCCAACGUU    7mer-m8\n",
       "2         hsa-miR-5680       17-23    AUUUCUA  CGAUUUCUAC    7mer-A1\n",
       "3       hsa-miR-379-5p       20-25     UCUACC  UUUCUACCUC       6mer\n",
       "4         hsa-miR-4451       20-25     UCUACC  UUUCUACCUC       6mer\n",
       "...                ...         ...        ...         ...        ...\n",
       "2302    hsa-miR-556-3p   2976-2982    GUAAUAA  GAGUAAUAAA    7mer-A1\n",
       "2303    hsa-miR-19a-5p   2991-2997    CAAAACA  GUCAAAACAG    7mer-A1\n",
       "2304  hsa-miR-19b-1-5p   2991-2997    CAAAACA  GUCAAAACAG    7mer-A1\n",
       "2305  hsa-miR-19b-2-5p   2991-2997    CAAAACA  GUCAAAACAG    7mer-A1\n",
       "2306      hsa-miR-2052   2991-2997    CAAAACA  GUCAAAACAG    7mer-A1\n",
       "\n",
       "[2307 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# run script\n",
    "fasta = \"../sequences/HMGA2_ENST00000403681_7_sequence.fa\"\n",
    "mirna_fasta = \"../sequences/mirna/mature_mirnas.fa\"\n",
    "\n",
    "s = import_fasta(fasta)\n",
    "m = create_mirna_matrix(mirna_fasta)\n",
    "\n",
    "df = find_matches(s, m)\n",
    "\n",
    "display(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "217\n"
     ]
    }
   ],
   "source": [
    "# results test to see how many of targetscan results are missing in our results\n",
    "\n",
    "df.to_excel(\"results.xlsx\")\n",
    "\n",
    "\n",
    "targetscan = \"TargetScan_8.0_ENST00000403681.2_predicted_targeting_details.xlsx\"\n",
    "ours = \"results.xlsx\"\n",
    "\n",
    "ts = pd.read_excel(targetscan)\n",
    "us = pd.read_excel(ours)\n",
    "\n",
    "ts_data = [ts[\"miRNA\"].to_list(), ts[\"Position in the UTR\"].to_list()]\n",
    "us_data = [us[\"name\"].to_list(), us[\"coordinates\"].to_list()]\n",
    "\n",
    "\n",
    "results = [[], []]\n",
    "missed_mirnas = [[], []]\n",
    "\n",
    "for record in ts_data[1]:\n",
    "\n",
    "    index_of_record = ts_data[1].index(record)\n",
    "\n",
    "    if record in us_data[1]:\n",
    "        results[0].append(ts_data[0][index_of_record])\n",
    "        results[1].append(record)\n",
    "\n",
    "    else:\n",
    "        missed_mirnas[0].append(ts_data[0][index_of_record])\n",
    "        missed_mirnas[1].append(record)\n",
    "\n",
    "\n",
    "print(len(missed_mirnas[0]))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "217 results of TargetScan are missing in our results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "200\n"
     ]
    }
   ],
   "source": [
    "missing_but_existing = []\n",
    "missing_and_not_existing = []\n",
    "\n",
    "for record in missed_mirnas[0]:\n",
    "    if record in m[0]:\n",
    "        missing_but_existing.append(record)\n",
    "\n",
    "    else:\n",
    "        missing_and_not_existing.append(record)\n",
    "\n",
    "\n",
    "print(len(missing_but_existing))\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "200 our of 217 missed records are exists in our miRNA database.\n",
    "\n",
    "17 of the 217 missing are miRNAs that doesn't exist in our miRNA database."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cells below are used for debugging, not a part of the script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# somewhat working\n",
    "\n",
    "\n",
    "def find_matches_debug(sequence, mirna_matrix):  # sourcery skip: remove-redundant-if\n",
    "\n",
    "    # mirna sequences are already converted to reverse complement (mRNA)\n",
    "    mirna_names = mirna_matrix[0]\n",
    "    mirna_sequences = mirna_matrix[1]\n",
    "\n",
    "    mirna_seeds = [seq[-7:-1] for seq in mirna_sequences]\n",
    "\n",
    "    seed_dict = defaultdict(list)\n",
    "\n",
    "    for i in range(len(mirna_names)):\n",
    "        seed_dict[mirna_seeds[i]].append((mirna_names[i], mirna_sequences[i]))\n",
    "\n",
    "    # generates k=10 length chunks of mRNA\n",
    "    generator = sliding_window(dna_to_mrna(sequence), 10)\n",
    "\n",
    "    names = []\n",
    "\n",
    "    m8_matches = []\n",
    "    seeds = []\n",
    "    a1s = []\n",
    "    chunks = []\n",
    "    match_types = []\n",
    "    mirnas_with_same_seed = []\n",
    "\n",
    "    # for loop\n",
    "    for c, chunk in enumerate(generator, start=1):\n",
    "\n",
    "        seed = chunk[2:8]\n",
    "\n",
    "        # if there's a seed match found\n",
    "        if seed in seed_dict:\n",
    "            for record in seed_dict[seed]:\n",
    "\n",
    "                mirna_sequence = seed_dict[seed][0][1]\n",
    "\n",
    "                if chunk[8] == \"A\" and chunk[1] == mirna_sequence[-8]:\n",
    "                    m8_matches.append(chunk[1])\n",
    "                    match_types.append(\"8mer\")\n",
    "\n",
    "                elif chunk[8] == \"A\":\n",
    "                    m8_matches.append(\"no\")\n",
    "                    match_types.append(\"7mer-A1\")\n",
    "\n",
    "                elif chunk[1] == mirna_sequence[-8]:\n",
    "                    m8_matches.append(chunk[1])\n",
    "                    match_types.append(\"7mer-m8\")\n",
    "\n",
    "                else:\n",
    "                    m8_matches.append(\"no\")\n",
    "                    match_types.append(\"6mer\")\n",
    "                mirnas_with_same_seed.append(len(seed_dict[seed]))\n",
    "\n",
    "                chunks.append(chunk)\n",
    "                a1s.append(chunk[8])\n",
    "\n",
    "                seeds.append(chunk[2:8])\n",
    "                names.append(record[0])\n",
    "\n",
    "    return (\n",
    "        pd.DataFrame(\n",
    "            list(\n",
    "                zip(\n",
    "                    names,\n",
    "                    m8_matches,\n",
    "                    seeds,\n",
    "                    a1s,\n",
    "                    chunks,\n",
    "                    match_types,\n",
    "                    mirnas_with_same_seed,\n",
    "                )\n",
    "            ),\n",
    "            columns=[\n",
    "                \"name\",\n",
    "                \"m8\",\n",
    "                \"seed\",\n",
    "                \"chunk[8]\",\n",
    "                \"chunk\",\n",
    "                \"match_type\",\n",
    "                \"mirnas with same seed\",\n",
    "            ],\n",
    "        ),\n",
    "        seed_dict,\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sequence loaded\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>m8</th>\n",
       "      <th>seed</th>\n",
       "      <th>chunk[8]</th>\n",
       "      <th>chunk</th>\n",
       "      <th>match_type</th>\n",
       "      <th>mirnas with same seed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-miR-6789-3p</td>\n",
       "      <td>G</td>\n",
       "      <td>GGCGCC</td>\n",
       "      <td>A</td>\n",
       "      <td>GGGGCGCCAA</td>\n",
       "      <td>8mer</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-miR-1306-3p</td>\n",
       "      <td>G</td>\n",
       "      <td>CCAACG</td>\n",
       "      <td>U</td>\n",
       "      <td>CGCCAACGUU</td>\n",
       "      <td>7mer-m8</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-miR-5680</td>\n",
       "      <td>no</td>\n",
       "      <td>AUUUCU</td>\n",
       "      <td>A</td>\n",
       "      <td>CGAUUUCUAC</td>\n",
       "      <td>7mer-A1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-miR-379-5p</td>\n",
       "      <td>no</td>\n",
       "      <td>UCUACC</td>\n",
       "      <td>U</td>\n",
       "      <td>UUUCUACCUC</td>\n",
       "      <td>6mer</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-miR-4451</td>\n",
       "      <td>no</td>\n",
       "      <td>UCUACC</td>\n",
       "      <td>U</td>\n",
       "      <td>UUUCUACCUC</td>\n",
       "      <td>6mer</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2302</th>\n",
       "      <td>hsa-miR-556-3p</td>\n",
       "      <td>no</td>\n",
       "      <td>GUAAUA</td>\n",
       "      <td>A</td>\n",
       "      <td>GAGUAAUAAA</td>\n",
       "      <td>7mer-A1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2303</th>\n",
       "      <td>hsa-miR-19a-5p</td>\n",
       "      <td>no</td>\n",
       "      <td>CAAAAC</td>\n",
       "      <td>A</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2304</th>\n",
       "      <td>hsa-miR-19b-1-5p</td>\n",
       "      <td>no</td>\n",
       "      <td>CAAAAC</td>\n",
       "      <td>A</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2305</th>\n",
       "      <td>hsa-miR-19b-2-5p</td>\n",
       "      <td>no</td>\n",
       "      <td>CAAAAC</td>\n",
       "      <td>A</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2306</th>\n",
       "      <td>hsa-miR-2052</td>\n",
       "      <td>no</td>\n",
       "      <td>CAAAAC</td>\n",
       "      <td>A</td>\n",
       "      <td>GUCAAAACAG</td>\n",
       "      <td>7mer-A1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2307 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  name  m8    seed chunk[8]       chunk match_type  \\\n",
       "0      hsa-miR-6789-3p   G  GGCGCC        A  GGGGCGCCAA       8mer   \n",
       "1      hsa-miR-1306-3p   G  CCAACG        U  CGCCAACGUU    7mer-m8   \n",
       "2         hsa-miR-5680  no  AUUUCU        A  CGAUUUCUAC    7mer-A1   \n",
       "3       hsa-miR-379-5p  no  UCUACC        U  UUUCUACCUC       6mer   \n",
       "4         hsa-miR-4451  no  UCUACC        U  UUUCUACCUC       6mer   \n",
       "...                ...  ..     ...      ...         ...        ...   \n",
       "2302    hsa-miR-556-3p  no  GUAAUA        A  GAGUAAUAAA    7mer-A1   \n",
       "2303    hsa-miR-19a-5p  no  CAAAAC        A  GUCAAAACAG    7mer-A1   \n",
       "2304  hsa-miR-19b-1-5p  no  CAAAAC        A  GUCAAAACAG    7mer-A1   \n",
       "2305  hsa-miR-19b-2-5p  no  CAAAAC        A  GUCAAAACAG    7mer-A1   \n",
       "2306      hsa-miR-2052  no  CAAAAC        A  GUCAAAACAG    7mer-A1   \n",
       "\n",
       "      mirnas with same seed  \n",
       "0                         1  \n",
       "1                         1  \n",
       "2                         1  \n",
       "3                         4  \n",
       "4                         4  \n",
       "...                     ...  \n",
       "2302                      1  \n",
       "2303                      4  \n",
       "2304                      4  \n",
       "2305                      4  \n",
       "2306                      4  \n",
       "\n",
       "[2307 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# debug run\n",
    "fasta = \"../sequences/HMGA2_ENST00000403681_7_sequence.fa\"\n",
    "mirna_fasta = \"../sequences/mirna/mature_mirnas.fa\"\n",
    "\n",
    "s = import_fasta(fasta)\n",
    "m = create_mirna_matrix(mirna_fasta)\n",
    "\n",
    "df, dict = find_matches_debug(s, m)\n",
    "\n",
    "display(df)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "tenv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "0a7314db4cd9cbf8825ff67208d8d351ac5340431b699125bf50d4fb6663edf3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
