{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/run/media/nazif/2F946E411BA61D49/thesis\n"
     ]
    }
   ],
   "source": [
    "cd .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /run/media/nazif/2F946E411BA61D49/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from pyensembl import EnsemblRelease\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option(\"display.max_columns\", None)\n",
    "\n",
    "\n",
    "# set pyensembl cache location\n",
    "os.environ['PYENSEMBL_CACHE_DIR'] = \"../data\"\n",
    "\n",
    "# init database\n",
    "g37 = EnsemblRelease(75)\n",
    "\n",
    "# download and index db\n",
    "g37.download()\n",
    "g37.index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Uploaded_variation</th>\n",
       "      <th>Allele</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>ENSP</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>TREMBL</th>\n",
       "      <th>UNIPARC</th>\n",
       "      <th>LOEUF</th>\n",
       "      <th>SpliceAI_pred_DP_AG</th>\n",
       "      <th>SpliceAI_pred_DP_AL</th>\n",
       "      <th>SpliceAI_pred_DP_DG</th>\n",
       "      <th>SpliceAI_pred_DP_DL</th>\n",
       "      <th>SpliceAI_pred_DS_AG</th>\n",
       "      <th>SpliceAI_pred_DS_AL</th>\n",
       "      <th>SpliceAI_pred_DS_DG</th>\n",
       "      <th>SpliceAI_pred_DS_DL</th>\n",
       "      <th>ada_score</th>\n",
       "      <th>rf_score</th>\n",
       "      <th>GO</th>\n",
       "      <th>CADD_PHRED</th>\n",
       "      <th>CADD_RAW</th>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <th>5_prime_UTR_variant</th>\n",
       "      <th>NMD_transcript_variant</th>\n",
       "      <th>TF_binding_site_variant</th>\n",
       "      <th>coding_sequence_variant</th>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <th>frameshift_variant</th>\n",
       "      <th>incomplete_terminal_codon_variant</th>\n",
       "      <th>intergenic_variant</th>\n",
       "      <th>intron_variant</th>\n",
       "      <th>mature_miRNA_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>non_coding_transcript_exon_variant</th>\n",
       "      <th>non_coding_transcript_variant</th>\n",
       "      <th>regulatory_region_variant</th>\n",
       "      <th>splice_acceptor_variant</th>\n",
       "      <th>splice_donor_5th_base_variant</th>\n",
       "      <th>splice_donor_region_variant</th>\n",
       "      <th>splice_donor_variant</th>\n",
       "      <th>splice_polypyrimidine_tract_variant</th>\n",
       "      <th>splice_region_variant</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>stop_lost</th>\n",
       "      <th>stop_retained_variant</th>\n",
       "      <th>synonymous_variant</th>\n",
       "      <th>upstream_gene_variant</th>\n",
       "      <th>HIGH</th>\n",
       "      <th>LOW</th>\n",
       "      <th>MODERATE</th>\n",
       "      <th>MODIFIER</th>\n",
       "      <th>MotifFeature</th>\n",
       "      <th>RegulatoryFeature</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>CTCF_binding_site</th>\n",
       "      <th>TF_binding_site</th>\n",
       "      <th>antisense</th>\n",
       "      <th>enhancer</th>\n",
       "      <th>lincRNA</th>\n",
       "      <th>miRNA</th>\n",
       "      <th>misc_RNA</th>\n",
       "      <th>nonsense_mediated_decay</th>\n",
       "      <th>open_chromatin_region</th>\n",
       "      <th>processed_pseudogene</th>\n",
       "      <th>processed_transcript</th>\n",
       "      <th>promoter</th>\n",
       "      <th>promoter_flanking_region</th>\n",
       "      <th>protein_coding</th>\n",
       "      <th>retained_intron</th>\n",
       "      <th>sense_intronic</th>\n",
       "      <th>snRNA</th>\n",
       "      <th>snoRNA</th>\n",
       "      <th>unprocessed_pseudogene</th>\n",
       "      <th>CTCF_interactions</th>\n",
       "      <th>CTCF_chains</th>\n",
       "      <th>POLR2A_interactions</th>\n",
       "      <th>POLR2A_chains</th>\n",
       "      <th>allele</th>\n",
       "      <th>3prime_overlapping_ncRNA</th>\n",
       "      <th>sense_overlapping</th>\n",
       "      <th>IG_C_gene</th>\n",
       "      <th>transcribed_unprocessed_pseudogene</th>\n",
       "      <th>rRNA</th>\n",
       "      <th>IG_J_gene</th>\n",
       "      <th>IG_J_pseudogene</th>\n",
       "      <th>non_stop_decay</th>\n",
       "      <th>start_lost</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>inframe_insertion</th>\n",
       "      <th>transcribed_processed_pseudogene</th>\n",
       "      <th>IG_D_gene</th>\n",
       "      <th>DNA</th>\n",
       "      <th>LINE</th>\n",
       "      <th>LTR</th>\n",
       "      <th>SINE</th>\n",
       "      <th>Simple_repeat</th>\n",
       "      <th>0kb</th>\n",
       "      <th>100kb downstream</th>\n",
       "      <th>100kb upstream</th>\n",
       "      <th>10kb downstream</th>\n",
       "      <th>10kb upstream</th>\n",
       "      <th>2kb downstream</th>\n",
       "      <th>2kb upstream</th>\n",
       "      <th>driver</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10_35318568_T/G</td>\n",
       "      <td>G</td>\n",
       "      <td>10</td>\n",
       "      <td>35318568</td>\n",
       "      <td>35318568</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>CUL2</td>\n",
       "      <td>-1</td>\n",
       "      <td>8</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0.093</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>T/G</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10_3827102_C/T</td>\n",
       "      <td>T</td>\n",
       "      <td>10</td>\n",
       "      <td>3827102</td>\n",
       "      <td>3827102</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>KLF6</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.252</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>196.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>198.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>C/T</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10_63661915_G/C</td>\n",
       "      <td>C</td>\n",
       "      <td>10</td>\n",
       "      <td>63661915</td>\n",
       "      <td>63661915</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>27</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>344.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>186.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>G/C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10_88681451_G/A</td>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>88681451</td>\n",
       "      <td>88681451</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>BMPR1A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.355</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>67</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>G/A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10_89653866_T/C</td>\n",
       "      <td>C</td>\n",
       "      <td>10</td>\n",
       "      <td>89653866</td>\n",
       "      <td>89653866</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0.507</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.999893</td>\n",
       "      <td>0.95</td>\n",
       "      <td>131</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>T/C</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #Uploaded_variation Allele chr     start       end ref alt  SYMBOL  STRAND  \\\n",
       "0     10_35318568_T/G      G  10  35318568  35318568   T   G    CUL2      -1   \n",
       "1      10_3827102_C/T      T  10   3827102   3827102   C   T    KLF6       0   \n",
       "2     10_63661915_G/C      C  10  63661915  63661915   G   C  ARID5B       1   \n",
       "3     10_88681451_G/A      A  10  88681451  88681451   G   A  BMPR1A       1   \n",
       "4     10_89653866_T/C      C  10  89653866  89653866   T   C    PTEN       1   \n",
       "\n",
       "   ENSP  SWISSPROT  TREMBL  UNIPARC  LOEUF  SpliceAI_pred_DP_AG  \\\n",
       "0     8          1       4        5  0.093                  0.0   \n",
       "1     3          1       1        3  0.252                  0.0   \n",
       "2     1          1       1        1  0.110                  0.0   \n",
       "3     1          1       1        1  0.355                  0.0   \n",
       "4     1          1       3        1  0.507                  0.0   \n",
       "\n",
       "   SpliceAI_pred_DP_AL  SpliceAI_pred_DP_DG  SpliceAI_pred_DP_DL  \\\n",
       "0                  0.0                  0.0                  0.0   \n",
       "1                  0.0                  0.0                  0.0   \n",
       "2                  0.0                  0.0                  0.0   \n",
       "3                  0.0                  0.0                  0.0   \n",
       "4                  0.0                  0.0                  0.0   \n",
       "\n",
       "   SpliceAI_pred_DS_AG  SpliceAI_pred_DS_AL  SpliceAI_pred_DS_DG  \\\n",
       "0                  0.0                  0.0                  0.0   \n",
       "1                  0.0                  0.0                  0.0   \n",
       "2                  0.0                  0.0                  0.0   \n",
       "3                  0.0                  0.0                  0.0   \n",
       "4                  0.0                  0.0                  0.0   \n",
       "\n",
       "   SpliceAI_pred_DS_DL  ada_score  rf_score   GO  CADD_PHRED  CADD_RAW  \\\n",
       "0                  0.0   0.000000      0.00   18         0.0       0.0   \n",
       "1                  0.0   0.000000      0.00    9         0.0       0.0   \n",
       "2                  0.0   0.000000      0.00   27         0.0       0.0   \n",
       "3                  0.0   0.000000      0.00   67         0.0       0.0   \n",
       "4                  0.0   0.999893      0.95  131         0.0       0.0   \n",
       "\n",
       "   3_prime_UTR_variant  5_prime_UTR_variant  NMD_transcript_variant  \\\n",
       "0                    1                    0                       1   \n",
       "1                    0                    0                       0   \n",
       "2                    0                    0                       0   \n",
       "3                    0                    0                       0   \n",
       "4                    0                    0                       0   \n",
       "\n",
       "   TF_binding_site_variant  coding_sequence_variant  downstream_gene_variant  \\\n",
       "0                        0                        0                        0   \n",
       "1                        0                        0                        0   \n",
       "2                        0                        0                        0   \n",
       "3                        0                        0                        0   \n",
       "4                        0                        0                        0   \n",
       "\n",
       "   frameshift_variant  incomplete_terminal_codon_variant  intergenic_variant  \\\n",
       "0                   0                                  0                   0   \n",
       "1                   0                                  0                   0   \n",
       "2                   0                                  0                   0   \n",
       "3                   0                                  0                   0   \n",
       "4                   0                                  0                   0   \n",
       "\n",
       "   intron_variant  mature_miRNA_variant  missense_variant  \\\n",
       "0               0                     0                 1   \n",
       "1               1                     0                 0   \n",
       "2               1                     0                 0   \n",
       "3               0                     0                 0   \n",
       "4               0                     0                 1   \n",
       "\n",
       "   non_coding_transcript_exon_variant  non_coding_transcript_variant  \\\n",
       "0                                   0                              0   \n",
       "1                                   0                              1   \n",
       "2                                   0                              0   \n",
       "3                                   0                              0   \n",
       "4                                   1                              0   \n",
       "\n",
       "   regulatory_region_variant  splice_acceptor_variant  \\\n",
       "0                          0                        0   \n",
       "1                          1                        0   \n",
       "2                          1                        0   \n",
       "3                          0                        0   \n",
       "4                          1                        0   \n",
       "\n",
       "   splice_donor_5th_base_variant  splice_donor_region_variant  \\\n",
       "0                              0                            0   \n",
       "1                              0                            1   \n",
       "2                              0                            0   \n",
       "3                              0                            0   \n",
       "4                              0                            0   \n",
       "\n",
       "   splice_donor_variant  splice_polypyrimidine_tract_variant  \\\n",
       "0                     0                                    0   \n",
       "1                     0                                    0   \n",
       "2                     0                                    1   \n",
       "3                     0                                    0   \n",
       "4                     0                                    0   \n",
       "\n",
       "   splice_region_variant  stop_gained  stop_lost  stop_retained_variant  \\\n",
       "0                      1            0          0                      0   \n",
       "1                      0            0          0                      0   \n",
       "2                      1            0          0                      0   \n",
       "3                      1            0          0                      0   \n",
       "4                      1            0          0                      0   \n",
       "\n",
       "   synonymous_variant  upstream_gene_variant  HIGH  LOW  MODERATE  MODIFIER  \\\n",
       "0                   0                      0     0    1         1         0   \n",
       "1                   0                      1     0    1         0         1   \n",
       "2                   0                      0     0    1         0         1   \n",
       "3                   1                      0     0    1         0         0   \n",
       "4                   0                      0     0    0         1         1   \n",
       "\n",
       "   MotifFeature  RegulatoryFeature  Transcript  CTCF_binding_site  \\\n",
       "0             0                  0           1                  0   \n",
       "1             0                  1           1                  0   \n",
       "2             0                  1           1                  0   \n",
       "3             0                  0           1                  0   \n",
       "4             0                  1           1                  0   \n",
       "\n",
       "   TF_binding_site  antisense  enhancer  lincRNA  miRNA  misc_RNA  \\\n",
       "0                0          0         0        0      0         0   \n",
       "1                0          0         0        0      0         0   \n",
       "2                0          0         0        0      0         0   \n",
       "3                0          0         0        0      0         0   \n",
       "4                0          0         0        0      0         0   \n",
       "\n",
       "   nonsense_mediated_decay  open_chromatin_region  processed_pseudogene  \\\n",
       "0                        1                      0                     0   \n",
       "1                        0                      0                     0   \n",
       "2                        0                      0                     0   \n",
       "3                        0                      0                     0   \n",
       "4                        0                      0                     0   \n",
       "\n",
       "   processed_transcript  promoter  promoter_flanking_region  protein_coding  \\\n",
       "0                     0         0                         0               1   \n",
       "1                     1         1                         0               1   \n",
       "2                     0         1                         0               1   \n",
       "3                     0         0                         0               1   \n",
       "4                     1         0                         1               1   \n",
       "\n",
       "   retained_intron  sense_intronic  snRNA  snoRNA  unprocessed_pseudogene  \\\n",
       "0                0               0      0       0                       0   \n",
       "1                0               0      0       0                       0   \n",
       "2                0               0      0       0                       0   \n",
       "3                0               0      0       0                       0   \n",
       "4                0               0      0       0                       0   \n",
       "\n",
       "   CTCF_interactions  CTCF_chains  POLR2A_interactions  POLR2A_chains allele  \\\n",
       "0               25.0         12.0                  8.0            3.0    T/G   \n",
       "1              196.0         66.0                198.0           86.0    C/T   \n",
       "2              344.0         86.0                186.0           74.0    G/C   \n",
       "3               31.0         10.0                  5.0            3.0    G/A   \n",
       "4                9.0          6.0                 11.0            4.0    T/C   \n",
       "\n",
       "   3prime_overlapping_ncRNA  sense_overlapping  IG_C_gene  \\\n",
       "0                         0                  0          0   \n",
       "1                         0                  0          0   \n",
       "2                         0                  0          0   \n",
       "3                         0                  0          0   \n",
       "4                         0                  0          0   \n",
       "\n",
       "   transcribed_unprocessed_pseudogene  rRNA  IG_J_gene  IG_J_pseudogene  \\\n",
       "0                                   0     0          0                0   \n",
       "1                                   0     0          0                0   \n",
       "2                                   0     0          0                0   \n",
       "3                                   0     0          0                0   \n",
       "4                                   0     0          0                0   \n",
       "\n",
       "   non_stop_decay  start_lost  pseudogene  inframe_insertion  \\\n",
       "0               0           0           0                  0   \n",
       "1               0           0           0                  0   \n",
       "2               0           0           0                  0   \n",
       "3               0           0           0                  0   \n",
       "4               0           0           0                  0   \n",
       "\n",
       "   transcribed_processed_pseudogene  IG_D_gene  DNA  LINE  LTR  SINE  \\\n",
       "0                                 0          0    0     0    0     0   \n",
       "1                                 0          0    0     0    0     0   \n",
       "2                                 0          0    0     0    0     0   \n",
       "3                                 0          0    0     0    0     0   \n",
       "4                                 0          0    0     0    0     0   \n",
       "\n",
       "   Simple_repeat  0kb  100kb downstream  100kb upstream  10kb downstream  \\\n",
       "0              0    0                 0               0                0   \n",
       "1              0    0                 0               0                0   \n",
       "2              0    0                 0               0                0   \n",
       "3              0    0                 0               0                0   \n",
       "4              0    0                 0               0                0   \n",
       "\n",
       "   10kb upstream  2kb downstream  2kb upstream  driver  \n",
       "0              0               0             0       1  \n",
       "1              0               0             0       1  \n",
       "2              0               0             0       1  \n",
       "3              0               0             0       1  \n",
       "4              0               0             0       1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(\"data/external/sana/final_dataset.csv\")\n",
    "uniques = df.SYMBOL.unique().tolist()\n",
    "\n",
    "def write_list_to_file(lst, filename):\n",
    "    with open(filename, 'w') as file:\n",
    "        for item in lst:\n",
    "            file.write(str(item) + '\\n')\n",
    "\n",
    "write_list_to_file(uniques, \"uniques.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "yes\n"
     ]
    }
   ],
   "source": [
    "if \"LINC00343\" in uniques:\n",
    "    print(\"yes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENST00000374742',\n",
       " 'ENST00000374746',\n",
       " 'ENST00000374748',\n",
       " 'ENST00000374749',\n",
       " 'ENST00000374751',\n",
       " 'ENST00000374754',\n",
       " 'ENST00000537177',\n",
       " 'ENST00000602371']"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g37.transcript_ids_at_locus(contig=10, position=35318568)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "bm = g37.genes_by_name(\"CUL2\")\n",
    "trans = bm[0].transcripts[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans.contains(start=35318568, end=35318568, contig=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(35360127, 35360245),\n",
       " (35351888, 35351990),\n",
       " (35349802, 35349896),\n",
       " (35343362, 35343467),\n",
       " (35338611, 35338693),\n",
       " (35333700, 35333796),\n",
       " (35333494, 35333604),\n",
       " (35327848, 35328010),\n",
       " (35324100, 35324224),\n",
       " (35322094, 35322201),\n",
       " (35321363, 35321422),\n",
       " (35320419, 35320547),\n",
       " (35320228, 35320314),\n",
       " (35318416, 35318568),\n",
       " (35317738, 35317815),\n",
       " (35314060, 35314126),\n",
       " (35305120, 35305322),\n",
       " (35302627, 35302728),\n",
       " (35299242, 35299370)]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans.coding_sequence_position_ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3903"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(bm[0].transcripts[0].sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "41699"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bm[0].transcripts[0].end - 35318568 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pyensembl.transcript.Transcript"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(bm[0].transcripts[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_names_at_locus(contig, position):\n",
    "    return g37.gene_names_at_locus(contig, position)\n",
    "\n",
    "def get_gene_ids_at_locus(contig, position):\n",
    "    return g37.genes_at_locus(contig, position)\n",
    "\n",
    "\n",
    "df[\"gene_name\"] = df.apply(lambda row: get_gene_names_at_locus(row[\"chr\"], row[\"start\"]), axis=1)\n",
    "df[\"gene_id\"] = df.apply(lambda row: get_gene_ids_at_locus(row[\"chr\"], row[\"start\"]), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name\n",
       "[]                       438\n",
       "[TP53]                    51\n",
       "[RFTN1]                   20\n",
       "[BZRAP1-AS1, MIR142]      14\n",
       "[NF1]                     10\n",
       "                        ... \n",
       "[LTB4R]                    1\n",
       "[AJUBA, RP11-298I3.5]      1\n",
       "[PARP2]                    1\n",
       "[PACS2]                    1\n",
       "[AKAP4]                    1\n",
       "Name: count, Length: 482, dtype: int64"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.gene_name.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Gene' object has no attribute 'coding_sequence'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[52], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m asd \u001b[39m=\u001b[39m g37\u001b[39m.\u001b[39mgenes_at_locus(\u001b[39m10\u001b[39m, \u001b[39m35318568\u001b[39m)\n\u001b[0;32m----> 3\u001b[0m asd[\u001b[39m0\u001b[39;49m]\u001b[39m.\u001b[39;49mcoding_sequence()\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Gene' object has no attribute 'coding_sequence'"
     ]
    }
   ],
   "source": [
    "asd = g37.genes_at_locus(10, 35318568)\n",
    "\n",
    "asd[0].coding_sequence()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000108094'"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = asd[0]\n",
    "\n",
    "gene.gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Allele</th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>new_col</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G</td>\n",
       "      <td>10</td>\n",
       "      <td>35318568</td>\n",
       "      <td>CUL2</td>\n",
       "      <td>[CUL2]</td>\n",
       "      <td>[CUL2]</td>\n",
       "      <td>[Gene(gene_id='ENSG00000108094', gene_name='CU...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>10</td>\n",
       "      <td>3827102</td>\n",
       "      <td>KLF6</td>\n",
       "      <td>[KLF6]</td>\n",
       "      <td>[KLF6]</td>\n",
       "      <td>[Gene(gene_id='ENSG00000067082', gene_name='KL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>10</td>\n",
       "      <td>63661915</td>\n",
       "      <td>ARID5B</td>\n",
       "      <td>[ARID5B]</td>\n",
       "      <td>[ARID5B]</td>\n",
       "      <td>[Gene(gene_id='ENSG00000150347', gene_name='AR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A</td>\n",
       "      <td>10</td>\n",
       "      <td>88681451</td>\n",
       "      <td>BMPR1A</td>\n",
       "      <td>[BMPR1A]</td>\n",
       "      <td>[BMPR1A]</td>\n",
       "      <td>[Gene(gene_id='ENSG00000107779', gene_name='BM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C</td>\n",
       "      <td>10</td>\n",
       "      <td>89653866</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>[PTEN]</td>\n",
       "      <td>[PTEN]</td>\n",
       "      <td>[Gene(gene_id='ENSG00000171862', gene_name='PT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Allele chr     start  SYMBOL   new_col gene_name  \\\n",
       "0      G  10  35318568    CUL2    [CUL2]    [CUL2]   \n",
       "1      T  10   3827102    KLF6    [KLF6]    [KLF6]   \n",
       "2      C  10  63661915  ARID5B  [ARID5B]  [ARID5B]   \n",
       "3      A  10  88681451  BMPR1A  [BMPR1A]  [BMPR1A]   \n",
       "4      C  10  89653866    PTEN    [PTEN]    [PTEN]   \n",
       "\n",
       "                                             gene_id  \n",
       "0  [Gene(gene_id='ENSG00000108094', gene_name='CU...  \n",
       "1  [Gene(gene_id='ENSG00000067082', gene_name='KL...  \n",
       "2  [Gene(gene_id='ENSG00000150347', gene_name='AR...  \n",
       "3  [Gene(gene_id='ENSG00000107779', gene_name='BM...  \n",
       "4  [Gene(gene_id='ENSG00000171862', gene_name='PT...  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "new_col\n",
       "[]                       438\n",
       "[TP53]                    51\n",
       "[RFTN1]                   20\n",
       "[BZRAP1-AS1, MIR142]      14\n",
       "[NF1]                     10\n",
       "                        ... \n",
       "[LTB4R]                    1\n",
       "[AJUBA, RP11-298I3.5]      1\n",
       "[PARP2]                    1\n",
       "[PACS2]                    1\n",
       "[AKAP4]                    1\n",
       "Name: count, Length: 482, dtype: int64"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.new_col.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CUL2']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g37.gene_names_at_locus(contig=10, position=35318568)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis-4DYweHFz-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
