{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>interaction ID</th>\n",
       "      <th>alignment score of target part</th>\n",
       "      <th>strand</th>\n",
       "      <th>miRNA ID</th>\n",
       "      <th>miRNA sequence</th>\n",
       "      <th>target sequence</th>\n",
       "      <th>number of reads</th>\n",
       "      <th>number of unique reads</th>\n",
       "      <th>read IDs</th>\n",
       "      <th>read sequences</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2</td>\n",
       "      <td>187627970</td>\n",
       "      <td>187627998</td>\n",
       "      <td>human08246</td>\n",
       "      <td>32</td>\n",
       "      <td>+</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AATGTTACTATTACGTCTTCTTCTGCCT</td>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>@ago2_kishore-hc_3834205_x1,@ago2_kishore-hc_4...</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTTTTATTACGTCTTCTTC,TGAGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>52382582</td>\n",
       "      <td>52382612</td>\n",
       "      <td>human04514</td>\n",
       "      <td>32</td>\n",
       "      <td>-</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AGAATCACACGTATGTGGTCTTCTGTGACT</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>@ago2_kishore-hc_1451499_x1</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGACGTATGTGGTGTTCTGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3</td>\n",
       "      <td>113530194</td>\n",
       "      <td>113530221</td>\n",
       "      <td>human11115</td>\n",
       "      <td>36</td>\n",
       "      <td>+</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AGGAAAGGTATTCTTCTATACAACTTG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>@ago2_kishore-pc_4275663_x1</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGACTCTTCTATACAACTTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM</td>\n",
       "      <td>14360</td>\n",
       "      <td>14382</td>\n",
       "      <td>human08122</td>\n",
       "      <td>34</td>\n",
       "      <td>+</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>CACAGCACCAATCCTACCTCCA</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>@ago2_kishore-hc_959688_x1</td>\n",
       "      <td>AGAGGTAGTAGGTTGTATAGCACCAATCCTACCTCCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13</td>\n",
       "      <td>29062126</td>\n",
       "      <td>29062156</td>\n",
       "      <td>human08660</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>CACGTTGGTCTCATGCCGCCTTCTGCTGGG</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>@ago2_kishore-hc_2120565_x1</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTCTCATACCGCCTTCTGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chromosome      start        end interaction ID  \\\n",
       "0       chr2  187627970  187627998     human08246   \n",
       "1      chr16   52382582   52382612     human04514   \n",
       "2       chr3  113530194  113530221     human11115   \n",
       "3       chrM      14360      14382     human08122   \n",
       "4      chr13   29062126   29062156     human08660   \n",
       "\n",
       "   alignment score of target part strand    miRNA ID          miRNA sequence  \\\n",
       "0                              32      +  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT   \n",
       "1                              32      -  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT   \n",
       "2                              36      +  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT   \n",
       "3                              34      +  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT   \n",
       "4                              31      -  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT   \n",
       "\n",
       "                  target sequence  number of reads  number of unique reads  \\\n",
       "0    AATGTTACTATTACGTCTTCTTCTGCCT                6                       6   \n",
       "1  AGAATCACACGTATGTGGTCTTCTGTGACT                1                       1   \n",
       "2     AGGAAAGGTATTCTTCTATACAACTTG                1                       1   \n",
       "3          CACAGCACCAATCCTACCTCCA                1                       1   \n",
       "4  CACGTTGGTCTCATGCCGCCTTCTGCTGGG                1                       1   \n",
       "\n",
       "                                            read IDs  \\\n",
       "0  @ago2_kishore-hc_3834205_x1,@ago2_kishore-hc_4...   \n",
       "1                        @ago2_kishore-hc_1451499_x1   \n",
       "2                        @ago2_kishore-pc_4275663_x1   \n",
       "3                         @ago2_kishore-hc_959688_x1   \n",
       "4                        @ago2_kishore-hc_2120565_x1   \n",
       "\n",
       "                                      read sequences  \n",
       "0  TGAGGTAGTAGGTTGTATAGTTTTATTACGTCTTCTTC,TGAGGTA...  \n",
       "1             TGAGGTAGTAGGTTGTATAGACGTATGTGGTGTTCTGT  \n",
       "2             TGAGGTAGTAGGTTGTATAGACTCTTCTATACAACTTG  \n",
       "3              AGAGGTAGTAGGTTGTATAGCACCAATCCTACCTCCA  \n",
       "4             TGAGGTAGTAGGTTGTATAGTCTCATACCGCCTTCTGC  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"grosswendt_raw.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>mirna_name</th>\n",
       "      <th>mirna_seq</th>\n",
       "      <th>target_seq</th>\n",
       "      <th>target_len</th>\n",
       "      <th>mirna_length</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr2_187627970_187627998_+_let-7a</td>\n",
       "      <td>chr2</td>\n",
       "      <td>187627970</td>\n",
       "      <td>187627998</td>\n",
       "      <td>+</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AATGTTACTATTACGTCTTCTTCTGCCT</td>\n",
       "      <td>28</td>\n",
       "      <td>22</td>\n",
       "      <td>grosswendt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr16_52382582_52382612_-_let-7a</td>\n",
       "      <td>chr16</td>\n",
       "      <td>52382582</td>\n",
       "      <td>52382612</td>\n",
       "      <td>-</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AGAATCACACGTATGTGGTCTTCTGTGACT</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>grosswendt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr3_113530194_113530221_+_let-7a</td>\n",
       "      <td>chr3</td>\n",
       "      <td>113530194</td>\n",
       "      <td>113530221</td>\n",
       "      <td>+</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>AGGAAAGGTATTCTTCTATACAACTTG</td>\n",
       "      <td>27</td>\n",
       "      <td>22</td>\n",
       "      <td>grosswendt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrM_14360_14382_+_let-7a</td>\n",
       "      <td>chrM</td>\n",
       "      <td>14360</td>\n",
       "      <td>14382</td>\n",
       "      <td>+</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>CACAGCACCAATCCTACCTCCA</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>grosswendt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr13_29062126_29062156_-_let-7a</td>\n",
       "      <td>chr13</td>\n",
       "      <td>29062126</td>\n",
       "      <td>29062156</td>\n",
       "      <td>-</td>\n",
       "      <td>hsa-let-7a</td>\n",
       "      <td>TGAGGTAGTAGGTTGTATAGTT</td>\n",
       "      <td>CACGTTGGTCTCATGCCGCCTTCTGCTGGG</td>\n",
       "      <td>30</td>\n",
       "      <td>22</td>\n",
       "      <td>grosswendt</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  id chromosome      start        end strand  \\\n",
       "0  chr2_187627970_187627998_+_let-7a       chr2  187627970  187627998      +   \n",
       "1   chr16_52382582_52382612_-_let-7a      chr16   52382582   52382612      -   \n",
       "2  chr3_113530194_113530221_+_let-7a       chr3  113530194  113530221      +   \n",
       "3          chrM_14360_14382_+_let-7a       chrM      14360      14382      +   \n",
       "4   chr13_29062126_29062156_-_let-7a      chr13   29062126   29062156      -   \n",
       "\n",
       "   mirna_name               mirna_seq                      target_seq  \\\n",
       "0  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT    AATGTTACTATTACGTCTTCTTCTGCCT   \n",
       "1  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT  AGAATCACACGTATGTGGTCTTCTGTGACT   \n",
       "2  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT     AGGAAAGGTATTCTTCTATACAACTTG   \n",
       "3  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT          CACAGCACCAATCCTACCTCCA   \n",
       "4  hsa-let-7a  TGAGGTAGTAGGTTGTATAGTT  CACGTTGGTCTCATGCCGCCTTCTGCTGGG   \n",
       "\n",
       "   target_len  mirna_length     dataset  \n",
       "0          28            22  grosswendt  \n",
       "1          30            22  grosswendt  \n",
       "2          27            22  grosswendt  \n",
       "3          22            22  grosswendt  \n",
       "4          30            22  grosswendt  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dropping irrelevant columns\n",
    "cols_to_drop = [\n",
    "    \"alignment score of target part\",\n",
    "    \"number of reads\",\n",
    "    \"number of unique reads\",\n",
    "    \"read IDs\",\n",
    "    \"read sequences\"\n",
    "]\n",
    "df = df.drop(columns=cols_to_drop)\n",
    "\n",
    "df[\"target_length\"] = df[\"end\"] - df[\"start\"]\n",
    "df[\"mirna_length\"] = df[\"miRNA sequence\"].apply(len)\n",
    "df[\"dataset\"] = \"grosswendt\"\n",
    "\n",
    "# adding id column\n",
    "df[\"id\"] = df[\"chromosome\"] + \"_\" + df[\"start\"].astype(str) + \"_\" + df[\"end\"].astype(str) + \"_\" + df[\"strand\"].astype(str) + \"_\" + df[\"miRNA ID\"].astype(str).str[4:]\n",
    "df.drop(columns=[\"interaction ID\"], inplace=True)\n",
    "\n",
    "rename_dict = {\n",
    "    \"chromosome\": \"chromosome\",\n",
    "    \"miRNA ID\": \"mirna_name\",\n",
    "    \"miRNA sequence\": \"mirna_seq\",\n",
    "    \"target sequence\": \"target_seq\",\n",
    "    \"target_length\": \"target_len\",\n",
    "    \"dataset\": \"dataset\"\n",
    "}\n",
    "\n",
    "df.rename(columns=rename_dict, inplace=True)\n",
    "\n",
    "# moving the id column to the start of the dataframe\n",
    "df = pd.concat([df.iloc[:, -1:], df.iloc[:, :-1]], axis=1)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv(\"../../processed/grosswendt/grosswendt.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.10"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
